<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • EnrichedHeatmap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EnrichedHeatmap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/EnrichedHeatmap.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/roadmap.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../articles/row_odering.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../articles/visualize_categorical_signals_wrapper.html">UNKNOWN TITLE</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/EnrichedHeatmap">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/EnrichedHeatmap/blob/master/vignettes/EnrichedHeatmap.Rmd"><code>vignettes/EnrichedHeatmap.Rmd</code></a></small>
      <div class="hidden name"><code>EnrichedHeatmap.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{1. Make Enriched Heatmaps}
-->
<div id="make-enriched-heatmaps" class="section level1">
<h1 class="hasAnchor">
<a href="#make-enriched-heatmaps" class="anchor"></a>Make Enriched Heatmaps</h1>
<p><strong>Author</strong>: Zuguang Gu ( <a href="mailto:z.gu@dkfz.de">z.gu@dkfz.de</a> )</p>
<p><strong>Date</strong>: 2019-09-29</p>
<hr>
<p>Enriched heatmap is a special type of heatmap which visualizes the enrichment of genomic signals over specific target regions. It is broadly used to visualize e.g. how histone modifications are enriched at transcription start sites.</p>
<p>There are already several tools that are able to make such heatmap (e.g. <a href="https://github.com/shenlab-sinai/ngsplot">ngs.plot</a>, <a href="https://github.com/fidelram/deepTools">deepTools</a> or <a href="https://bioconductor.org/packages/release/bioc/html/genomation.html">genomation</a>). Here we implement enriched heatmap by <strong>ComplexHeatmap</strong> package. Since the enriched heatmap is essentially a normal heatmap but with some special settings, with the functionality of <strong>ComplexHeatmap</strong>, it would be much easier to customize the heatmaps as well as concatenating to a list of heatmaps to show correspondance between different data sources.</p>
<div id="basic-usages" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usages" class="anchor"></a>Basic usages</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(EnrichedHeatmap)</code></pre></div>
<p>First let’s load the example dataset that we will use for demonstration. The data for human lung tissue is from <a href="http://www.roadmapepigenomics.org/data/">Roadmap dataset</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">123</span>)
<span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"chr21_test_data.RData"</span>, <span class="dt">package =</span> <span class="st">"EnrichedHeatmap"</span>))
<span class="kw"><a href="https://rdrr.io/r/base/ls.html">ls</a></span>()</code></pre></div>
<pre><code>## [1] "cgi"     "genes"   "H3K4me3" "meth"    "rpkm"</code></pre>
<p>There are following R objects:</p>
<ul>
<li>
<code>H3K4me3</code>: coverage for H3K4me3 histone modification from the ChIP-seq data,</li>
<li>
<code>cgi</code>: CpG islands,</li>
<li>
<code>genes</code>: genes,</li>
<li>
<code>meth</code>: methylation for CpG sites from WGBS,</li>
<li>
<code>rpkm</code>: gene expression from RNASeq.</li>
</ul>
<p>In order to build the vignette fast, the data only includes chromosome 21. Also we downsampled 100000 CpG sites for methylation data.</p>
<p>We first visualize how H3K4me3 histone modification is enriched around gene TSS. We extract TSS of genes (note <code>tss</code> has strand information):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tss =<span class="st"> </span><span class="kw">promoters</span>(genes, <span class="dt">upstream =</span> <span class="dv">0</span>, <span class="dt">downstream =</span> <span class="dv">1</span>)
tss[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<pre><code>## GRanges object with 5 ranges and 0 metadata columns:
##                      seqnames    ranges strand
##                         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    ENSG00000141956.9    chr21  43299591      -
##   ENSG00000141959.12    chr21  45719934      +
##    ENSG00000142149.4    chr21  33245628      +
##   ENSG00000142156.10    chr21  47401651      +
##    ENSG00000142166.8    chr21  34696734      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">H3K4me3[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<pre><code>## GRanges object with 5 ranges and 1 metadata column:
##       seqnames          ranges strand |  coverage
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]    chr21 9825468-9825470      * |        10
##   [2]    chr21 9825471-9825488      * |        13
##   [3]    chr21         9825489      * |        12
##   [4]    chr21         9825490      * |        13
##   [5]    chr21 9825491-9825493      * |        14
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Similar as other tools, the task of visualization are separated into two steps:</p>
<ol style="list-style-type: decimal">
<li>get association between genomic signals and targets by normalizing into a matrix.</li>
<li>visualize the matrix by heatmap.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>)
mat1</code></pre></div>
<pre><code>## Normalize H3K4me3 to tss:
##   Upstream 5000 bp (100 windows)
##   Downstream 5000 bp (100 windows)
##   Include target regions (width = 1)
##   720 target regions</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(mat1)</code></pre></div>
<pre><code>## [1] "normalizedMatrix" "matrix"</code></pre>
<p><code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix()</a></code> converts the association between genomic signals (<code>H3K4me3</code>) and targets(<code>tss</code>) into a matrix (actually <code>mat1</code> is just a normal matrix with several additional attributes). It first splits the extended targets regions (the extension to upstream and downstream is controlled by <code>extend</code> argument) into a list of small windows (the width of the windows is controlled by <code>w</code>), then overlaps genomic signals to these small windows and calculates the value for every small window which is the mean value of genomic signals that intersects with the window (the value corresponds to genomic signals are controlled by <code>value_column</code> and how to calcualte the mean value is controlled by <code>mean_mode</code>.).</p>
<p>There are several modes for <code>mean_mode</code> according to different types of genomic signals. It will be explained in later sections.</p>
<p>With <code>mat1</code>, we can visualize it as a heatmap:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-7-1.png" width="288" style="display: block; margin: auto;"></p>
<p>By default, rows are ordered according to the enrichment to the target regions. On top of the heatmap there is a specific type of annotation which summarize the enrichment patterns as a line plot, implemented b <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched()</a></code>.</p>
<p><code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap()</a></code> internally calls <code>Heatmap()</code> and returns a <code>Heatmap</code> class object, so parameters for <code>Heatmap()</code> can be directly applied to <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap()</a></code>. Users can go to the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/"><strong>ComplexHeatmap</strong> package</a> to get a more comprehensive help.</p>
<div id="colors" class="section level3">
<h3 class="hasAnchor">
<a href="#colors" class="anchor"></a>Colors</h3>
<p>Similar as the normal heatmap, the simplest way to set colors is to provide a vector of colors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>), <span class="dt">name =</span> <span class="st">"H3K4me3"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-8-1.png" width="288" style="display: block; margin: auto;"></p>
<p>You may wonder why the color looks so light. The reason is in coverage values in <code>H3K4me3</code>, there exist some extreme values, which results in extreme value in <code>mat1</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(H3K4me3<span class="op">$</span>coverage, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">0.99</span>, <span class="dv">1</span>))</code></pre></div>
<pre><code>##   0%  25%  50%  75%  99% 100% 
##   10   18   29   43   87  293</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(mat1, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">0.99</span>, <span class="dv">1</span>))</code></pre></div>
<pre><code>##        0%       25%       50%       75%       99%      100% 
##   0.00000   0.00000   0.00000   0.00000  38.92176 174.78000</code></pre>
<p>If a vector of colors is specified, sequential values from minimal to maximal are mapped to the colors, and other values are linearly interpolated. To get rid of such extreme values, there are two ways. The first is to specify <code>keep</code> option which trims extreme values both at lower and upper bounds. (In following, it means only to trim values larger than 99th percentile.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat1_trim =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">keep =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.99</span>))
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1_trim, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>), <span class="dt">name =</span> <span class="st">"H3K4me3"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-10-1.png" width="288" style="display: block; margin: auto;"></p>
<p>The second way is to define a color mapping function which only maps colors to values less than 99th percentile and the value larger than the 99th percentile uses same color as the 99th percentile.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(circlize)
col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(mat1, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.99</span>)), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>))
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-11-1.png" width="288" style="display: block; margin: auto;"></p>
<p>To sum it up, the first way directly modified values in <code>mat1</code> while the second way keeps the original values but uses a robust color mapping.</p>
<p>If <code>col</code> is not specified in <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap()</a></code>, blue-white-red is mapped to 1st quantile, mean and 99th quantile by default.</p>
<p>In following sections, we will also use the matrix to do row-clustering, thus we directly use the trimmed matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat1 =<span class="st"> </span>mat1_trim</code></pre></div>
</div>
<div id="split-on-rows" class="section level3">
<h3 class="hasAnchor">
<a href="#split-on-rows" class="anchor"></a>Split on rows</h3>
<p>Split rows by a vector or a data frame by specifying <code>row_split</code> option.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, 
    <span class="dt">row_split =</span> <span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(genes), <span class="dt">replace =</span> <span class="ot">TRUE</span>),
    <span class="dt">column_title =</span> <span class="st">"Enrichment of H3K4me3"</span>) </code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-13-1.png" width="288" style="display: block; margin: auto;"></p>
<p>Split rows by k-means clustering by specifying <code>row_km</code> option.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">123</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">row_km =</span> <span class="dv">3</span>,
    <span class="dt">column_title =</span> <span class="st">"Enrichment of H3K4me3"</span>, <span class="dt">row_title_rot =</span> <span class="dv">0</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-14-1.png" width="384" style="display: block; margin: auto;"></p>
<p>In each row cluster, rows are still ordered by the enrichment.</p>
<p>When rows are split, graphic parameters for the enriched annotation can be a vector with length as the number of row clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">123</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">row_km =</span> <span class="dv">3</span>,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enriched =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>))),
    <span class="dt">column_title =</span> <span class="st">"Enrichment of H3K4me3"</span>, <span class="dt">row_title_rot =</span> <span class="dv">0</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-15-1.png" width="384" style="display: block; margin: auto;"></p>
<p>Users can go to <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-split" class="uri">https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-split</a> for more controls of splitting heatmaps.</p>
</div>
<div id="cluster-on-rows" class="section level3">
<h3 class="hasAnchor">
<a href="#cluster-on-rows" class="anchor"></a>Cluster on rows</h3>
<p>Cluster on rows. By default <code>show_row_dend</code> is turned off, so you don’t need to specify it here. More options for row clustering can be found in the help page of <code>Heatmap()</code> or <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#clustering" class="uri">https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#clustering</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, 
    <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>, <span class="dt">column_title =</span> <span class="st">"Enrichment of H3K4me3"</span>)     </code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-16-1.png" width="288" style="display: block; margin: auto;"></p>
<p>Vignette <a href="row_ordering.html" class="uri">row_ordering.html</a> compares different row ordering methods and clustering methods, and discusses which might be the proper way to show the enrichment patterns.</p>
</div>
<div id="extensions-to-target-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#extensions-to-target-regions" class="anchor"></a>Extensions to target regions</h3>
<p>Extension to upstream and downstream can be controled by <code>extend</code> either by a single value or a vector of length 2.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># upstream 1kb, downstream 2kb</span>
mat12 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1000</span>, <span class="dv">2000</span>), <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat12, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">col =</span> col_fun)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-17-1.png" width="288" style="display: block; margin: auto;"></p>
<p>Either upstream or downstream can be set to 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat12 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">2000</span>), <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat12, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">col =</span> col_fun)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-18-1.png" width="288" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat12 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1000</span>, <span class="dv">0</span>), <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat12, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">col =</span> col_fun)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-18-2.png" width="288" style="display: block; margin: auto;"></p>
<p>Sometimes whether the signals are on the upstream or the downstream of the targets are not important and users only want to show the relative distance to targets. If <code>flip_upstream</code> is set to <code>TRUE</code>, the upstream part in the normalized matrix is flipped and added to the downstream part. The flipping is only allowed when the targets are single-point targets or the targets are excluded in the normalized matrix (by setting <code>include_target = FALSE</code>). If the extension for the upstream and downstream is not equal, the smaller extension is used for the final matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_f =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">flip_upstream =</span> <span class="ot">TRUE</span>)
mat_f</code></pre></div>
<pre><code>## Normalize H3K4me3 to tss:
##   Extension 5000 bp (100 window)
##     upstream is flipped to downstream.
##   Include target regions (width = 1)
##   720 target regions</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_f, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">col =</span> col_fun)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-19-1.png" width="288" style="display: block; margin: auto;"></p>
</div>
<div id="axis-of-the-enriched-annotation" class="section level3">
<h3 class="hasAnchor">
<a href="#axis-of-the-enriched-annotation" class="anchor"></a>Axis of the enriched annotation</h3>
<p>Axis of the enriched annotation is controlled by <code>axis_param</code> in <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched()</a></code>. All the parameters that can be set can be found in <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/default_axis_param.html">ComplexHeatmap::default_axis_param()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, 
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(
        <span class="dt">enriched =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(
            <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">10</span>),
            <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
                <span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">5</span>, <span class="dv">10</span>),
                <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"zero"</span>, <span class="st">"five"</span>, <span class="st">"ten"</span>),
                <span class="dt">side =</span> <span class="st">"left"</span>,
                <span class="dt">facing =</span> <span class="st">"outside"</span>
    )))
)     </code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-20-1.png" width="288" style="display: block; margin: auto;"></p>
</div>
<div id="mean-mode" class="section level3">
<h3 class="hasAnchor">
<a href="#mean-mode" class="anchor"></a>Mean mode</h3>
<p>When normalizing genomic signals to target regions, upstream and downstream (also target regions themselves if they are included) of the targets are split into small windows. Then genomic signals are overlapped to each window and mean signal for each window is calculated. When a window is not completely covered by the regions for the genomic signales, proper averaging method should be applied to summarize the value in the window.</p>
<p>Depending on different scenarios, <strong>EnrichedHeatmap</strong> provides three metrics for averaging.</p>
<p>The overlapping model is illustrated in the following plot. The red line in the bottom represents the small window. Black lines on the top are the regions for genomic signals that overlap with the window. The thick lines indicate the intersected part between the signal regions and the window.</p>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-21-1.png" width="700" style="display: block; margin: auto;"></p>
<p>For a given window, <span class="math inline">\(n\)</span> is the number of signal regions which overlap with the window (it is 5 in the above plot), <span class="math inline">\(w_i\)</span> is the width of the intersected segments (black thick lines), and <span class="math inline">\(x_i\)</span> is the signal value associated with the original regions. If there is no value associated with the signal regions, <span class="math inline">\(x_i = 1\)</span> by default.</p>
<p>The “absolute” method is denoted as <span class="math inline">\(v_a\)</span> and is simply calculated as the mean of all signal regions regardless of their width:</p>
<p><span class="math display">\[ v_a = \frac{\sum_i^n{x_i}}{n} \]</span></p>
<p>The “weighted” method is denoted as <span class="math inline">\(v_w\)</span> and is calculated as the mean of all signal regions weighted by the width of their intersections:</p>
<p><span class="math display">\[ v_w = \frac{\sum_i^n{x_iw_i}}{\sum_i^n{w_i}} \]</span></p>
<p>“Absolute” and “weighted” methods should be applied when background values should not be taken into consideration. For example, when summarizing the mean methylation in a small window, non-CpG background should be ignored, because methylation is only associated with CpG sites and not with other positions.</p>
<p>The “w0” method is the weighted mean between the intersected parts and un-intersected parts:</p>
<p><span class="math display">\[ v_{w0} = \frac{v_wW}{W+W'} \]</span></p>
<p><span class="math inline">\(W\)</span> is sum of width of the intersected parts (<span class="math inline">\(\sum_i^n{w_i}\)</span>) and <span class="math inline">\(W'\)</span> is the sum of width for the non-intersected parts.</p>
<p>The “coverage” method is denoted as <span class="math inline">\(v_c\)</span> and is defined as the mean signal averged by the length of the window:</p>
<p><span class="math display">\[ v_c = \frac{\sum_i^n{x_iw_i}}{L} \]</span></p>
<p>where <span class="math inline">\(L\)</span> is the length of the window itself. Note when <span class="math inline">\(x_i = 1\)</span>, <span class="math inline">\(v_c\)</span> is the mean coverage for the signal regions overlapped in the window.</p>
<p>Following illustrates different settings for <code>mean_mode</code> (note there is a signal region overlapping with other signal regions):</p>
<pre><code>   40      50     20     values in signal regions
 ++++++   +++    +++++   signal regions
        30               values in signal regions
      ++++++             signal regions
   =================     window (17bp), there are 4bp not overlapping to any signal region.
     4  6  3      3      overlap

 absolute: (40 + 30 + 50 + 20)/4
 weighted: (40*4 + 30*6 + 50*3 + 20*3)/(4 + 6 + 3 + 3)
 w0:       (40*4 + 30*6 + 50*3 + 20*3)/(4 + 6 + 3 + 3 + 4)
 coverage: (40*4 + 30*6 + 50*3 + 20*3)/17</code></pre>
</div>
<div id="smoothing" class="section level3">
<h3 class="hasAnchor">
<a href="#smoothing" class="anchor"></a>Smoothing</h3>
<p>Rows can be smoothed by setting <code>smooth</code> to <code>TRUE</code> when generating the matrix. Later we will demonstrate smoothing can also help to impute <code>NA</code> values.</p>
<p>As smoothing may change the original data range, the color mapping function <code>col_fun</code> here ensures that the color palette is still the same as the unsmoothed one.</p>
<p><code>background</code> corresponds to the regions that have no signal overlapped. The proper value depends on specific scenarios. Here since we visualize coverage from ChIP-Seq data, it is reasonable to assign 0 to regions with no H3K4me3 signal.</p>
<p>In following example, since a enriched heatmap is also a heatmap, we can concatenate two heamtaps by <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat1_smoothed =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">background =</span> <span class="dv">0</span>, <span class="dt">smooth =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1_smoothed, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3_smoothed"</span>,
    <span class="dt">column_title =</span> <span class="st">"smoothed"</span>) <span class="op">+</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>, <span class="dt">column_title =</span> <span class="st">"unsmoothed"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/smooth-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In above plots, you might feel the left heatmap is better than the right unsmoothed heatmap. In following, we will demonstrate smoothing can significantly improve the enrichment pattern for methylation datasets.</p>
<p>Following heatmap visualizes the enrichment of low methylated regions over TSS. The grey colors represent the windows with no CpG sites (note we set <code>NA</code> to <code>background</code> and grey is the default color for <code>NA</code> values by <strong>ComplexHeatmap</strong>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<pre><code>## GRanges object with 5 ranges and 1 metadata column:
##       seqnames    ranges strand |              meth
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]    chr21   9432427      * | 0.267104203931852
##   [2]    chr21   9432428      * | 0.267106771955287
##   [3]    chr21   9432964      * | 0.272709911227985
##   [4]    chr21   9432965      * |   0.2727345344132
##   [5]    chr21   9433315      * | 0.285114797969136
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, tss, <span class="dt">value_column =</span> <span class="st">"meth"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">background =</span> <span class="ot">NA</span>)
meth_col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat2, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>, <span class="dt">column_title =</span> <span class="st">"methylation near TSS"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-22-1.png" width="288" style="display: block; margin: auto;"></p>
<p>When overlapping CpG positions to segmented target regions, it is possible that there is no CpG sites in some windows, especially for <code>meth</code> which only contains 100000 CpG sites which are randomly sampled in chromosome 21. The values for these windows which contain no CpG sites can be imputed by smoothing. Although it seems not proper to assign methylation values to non- CpG windows, but it will improve the visualization a lot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, tss, <span class="dt">value_column =</span> <span class="st">"meth"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">background =</span> <span class="ot">NA</span>, <span class="dt">smooth =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat2, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>, <span class="dt">column_title =</span> <span class="st">"methylation near TSS"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-23-1.png" width="288" style="display: block; margin: auto;"></p>
<p>To do the smoothing, by default, <code>locfit()</code> is first applied to each row in the original matrix. If it is failed, <code><a href="https://rdrr.io/r/stats/loess.html">loess()</a></code> smoothing is applied afterwards. If both smoothing methods are failed, there will be a warning and the original value is kept.</p>
<p>Users can provides their own smoothing function by <code>smooth_fun</code> argument. This self-defined function accepts a numeric vector (may contains <code>NA</code> values) and returns a vector with same length. If the smoothing is failed, the function should call <code><a href="https://rdrr.io/r/base/stop.html">stop()</a></code> to throw errors so that <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix()</a></code> can catch how many rows are failed in smoothing. Take a look at the source code of <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/default_smooth_fun.html">default_smooth_fun()</a></code> to get an example.</p>
</div>
<div id="targets-are-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#targets-are-regions" class="anchor"></a>Targets are regions</h3>
<p>In the example of H3K4me3, the target regions are single points. The targets can also be regions with width larger than 1. Following heatmap visualizes the enrichment of low methylation on CpG islands:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, cgi, <span class="dt">value_column =</span> <span class="st">"meth"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">background =</span> <span class="ot">NA</span>, <span class="dt">smooth =</span> <span class="ot">TRUE</span>, <span class="dt">target_ratio =</span> <span class="fl">0.3</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat3, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>, <span class="dt">axis_name_rot =</span> <span class="dv">90</span>,
    <span class="dt">column_title =</span> <span class="st">"methylation near CGI"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-24-1.png" width="288" style="display: block; margin: auto;"></p>
<p>Width of the target regions shown on heatmap can be controlled by <code>target_ratio</code> which is relative to the width of the complete heatmap.</p>
<p>Target regions are also splitted into small windows. Due to the unequal width of target regions, each target is split into <span class="math inline">\(k\)</span> equal windows with <span class="math inline">\(k = (n_1 +n_2)*r/(1-r)\)</span> where <span class="math inline">\(n_1\)</span> is the number of upstream windows, <span class="math inline">\(n_2\)</span> is the number of downstream windows and <span class="math inline">\(r\)</span> is the ratio of target columns presented in the matrix. There is a <code>k</code> argument in <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix()</a></code>, but it is only used when there is no upstream nor downstream for the targets.</p>
<p>When genomic targets are regions, upstream and/or downstream can be excluded in the heatmap.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, cgi, <span class="dt">value_column =</span> <span class="st">"meth"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">5000</span>), <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">background =</span> <span class="ot">NA</span>, <span class="dt">smooth =</span> <span class="ot">TRUE</span>, <span class="dt">target_ratio =</span> <span class="fl">0.5</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat3, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>, 
    <span class="dt">column_title =</span> <span class="st">"methylation near CGI"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/extension-1.png" width="288" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, cgi, <span class="dt">value_column =</span> <span class="st">"meth"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5000</span>, <span class="dv">0</span>), <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">background =</span> <span class="ot">NA</span>, <span class="dt">smooth =</span> <span class="ot">TRUE</span>, <span class="dt">target_ratio =</span> <span class="fl">0.5</span>)</code></pre></div>
<pre><code>## Warning: Smoothing is failed for one row because there are very few signals overlapped to it.
## Please use `failed_rows(mat)` to get the index of the failed row and consider to remove
## it.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat3, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>, 
    <span class="dt">column_title =</span> <span class="st">"methylation near CGI"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/extension-2.png" width="288" style="display: block; margin: auto;"></p>
<p>When there is no upstream nor downstream, the number of columns in the heatmap is controlled by <code>k</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, cgi, <span class="dt">value_column =</span> <span class="st">"meth"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">extend =</span> <span class="dv">0</span>, <span class="dt">k =</span> <span class="dv">20</span>, <span class="dt">background =</span> <span class="ot">NA</span>, <span class="dt">smooth =</span> <span class="ot">TRUE</span>, <span class="dt">target_ratio =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>## Warning: Smoothing is failed for one row because there are very few signals overlapped to it.
## Please use `failed_rows(mat)` to get the index of the failed row and consider to remove
## it.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat3, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>,
    <span class="dt">column_title =</span> <span class="st">"methylation near CGI"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-25-1.png" width="288" style="display: block; margin: auto;"></p>
<p>You may notice there are warnings when executing above code, that is because there are very few signals overlapped to some rows, which results too many <code>NA</code> values and failed with the smoothing. Corresponding index for failed rows can be get by :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/failed_rows.html">failed_rows</a></span>(mat3)</code></pre></div>
<pre><code>## [1] 5</code></pre>
<p>and maybe you can remove this row in the matrix beforehand.</p>
</div>
</div>
<div id="multiple-heatmaps" class="section level2">
<h2 class="hasAnchor">
<a href="#multiple-heatmaps" class="anchor"></a>Multiple heatmaps</h2>
<p>The power of <strong>EnrichedHeatmap</strong> package is that parallel heatmaps can be concatenated, both for enriched heatmap, normal heatmap as well the row annotations, which provides a very efficient way to visualize multiple sources of information.</p>
<p>With the functionality of <strong>ComplexHeatmap</strong> package, heatmaps can be concatenated by <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> operator. <code>Heatmap</code> objects and <code>HeatmapAnnotation</code> objects can be mixed.</p>
<p>Following heatmaps visualizes correspondance between H3K4me3 modification, methylation and gene expression. It is quite straightforward to see high expression correlates with low methylation and high H3K4me3 signal around TSS.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"left"</span>)))) <span class="op">+</span><span class="st"> </span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat2, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>) <span class="op">+</span>
<span class="kw">Heatmap</span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(rpkm<span class="op">+</span><span class="dv">1</span>), <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"orange"</span>), <span class="dt">name =</span> <span class="st">"log2(rpkm+1)"</span>, 
    <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">"mm"</span>))</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-27-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Of course you can split rows by partition variables or k-means clustering in the main heatmap. In following heatmaps, the most right color bar can be corresponded to the colors in column annotation on both histone modification heatmap and methylation heatmap.</p>
<p>Here we emphasize again, proper trimming on the matrix will greatly help to reveal the patterns. You can try replace <code>mat1</code> to a un-trimmed matrix and see whether this patterns shown below still preserves.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">partition =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"cluster"</span>, <span class="kw"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(mat1, <span class="dt">centers =</span> <span class="dv">3</span>)<span class="op">$</span>cluster)
lgd =<span class="st"> </span><span class="kw">Legend</span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cluster1"</span>, <span class="st">"cluster2"</span>, <span class="st">"cluster3"</span>), <span class="dt">title =</span> <span class="st">"Clusters"</span>, 
    <span class="dt">type =</span> <span class="st">"lines"</span>, <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>))
ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(partition, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/structure.html">structure</a></span>(<span class="dv">2</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">names =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"cluster"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)), <span class="dt">name =</span> <span class="st">"partition"</span>,
              <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">3</span>, <span class="st">"mm"</span>)) <span class="op">+</span>
<span class="st">          </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">"H3K4me3"</span>,
              <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>))), 
              <span class="dt">column_title =</span> <span class="st">"H3K4me3"</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">          </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat2, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"methylation"</span>,
              <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>))), 
              <span class="dt">column_title =</span> <span class="st">"Methylation"</span>) <span class="op">+</span>
<span class="st">          </span><span class="kw">Heatmap</span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(rpkm<span class="op">+</span><span class="dv">1</span>), <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"orange"</span>), <span class="dt">name =</span> <span class="st">"log2(rpkm+1)"</span>, 
              <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">15</span>, <span class="st">"mm"</span>),
              <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">summary =</span> <span class="kw">anno_summary</span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>),
                <span class="dt">outline =</span> <span class="ot">FALSE</span>, <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>))))
<span class="kw">draw</span>(ht_list, <span class="dt">split =</span> partition, <span class="dt">annotation_legend_list =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(lgd), 
    <span class="dt">ht_gap =</span> <span class="kw">unit</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">8</span>, <span class="dv">8</span>), <span class="st">"mm"</span>))</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="visualize-positive-signals-and-negative-signals-separatedly" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-positive-signals-and-negative-signals-separatedly" class="anchor"></a>Visualize positive signals and negative signals separatedly</h2>
<p>Sometimes we visualize the general correlation or the group difference around certain genomic targets. In this case, it makes more sense to visualize the enrichment for the positive signals and negative signals separatedly. In following example, variable <code>mat_H3K4me1</code> contains correlation between H3K4me1 signal and gene expression in (-5kb, 10kb) of the gene TSS.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"H3K4me1_corr_normalize_to_tss.RData"</span>, <span class="dt">package =</span> <span class="st">"EnrichedHeatmap"</span>))
mat_H3K4me1</code></pre></div>
<pre><code>## Normalize  to target:
##   Upstream 5000 bp (100 windows)
##   Downstream 10000 bp (200 windows)
##   Include target regions (width = 1)
##   677 target regions</code></pre>
<p>In <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched()</a></code>, there are two non-standard parameters <code>neg_col</code> and <code>pos_col</code> for <code>gp</code>. If these two are set, the enrichment lines are drawn for the positive signals and negative signals in the matrix separatedly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">corr_col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"darkgreen"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_H3K4me1, <span class="dt">col =</span> corr_col_fun, <span class="dt">name =</span> <span class="st">"corr_H3K4me1"</span>,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(
        <span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">neg_col =</span> <span class="st">"darkgreen"</span>, <span class="dt">pos_col =</span> <span class="st">"red"</span>),
            <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"left"</span>))
    ), <span class="dt">column_title =</span> <span class="st">"separate neg and pos"</span>) <span class="op">+</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_H3K4me1, <span class="dt">col =</span> corr_col_fun, <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">value =</span> <span class="st">"abs_mean"</span>)),
    <span class="dt">column_title =</span> <span class="st">"pool neg and pos"</span>)</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-30-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="restrict-overlapping-by-mappings" class="section level2">
<h2 class="hasAnchor">
<a href="#restrict-overlapping-by-mappings" class="anchor"></a>Restrict overlapping by mappings</h2>
<p>By default every genomic signal tries to intersect to every target region, but if mapping is provided, only those genomic signals that are mapped to the corresponding target region will be overlapped.</p>
<p>To illustrate it more clearly, we load the example data. <code>gene</code> column in <code>neg_cr</code> is used to map to the names of <code>all_tss</code>. In following example, <code>neg_cr</code> is the signal and <code>all_tss</code> is the target.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"neg_cr.RData"</span>, <span class="dt">package =</span> <span class="st">"EnrichedHeatmap"</span>))
all_tss =<span class="st"> </span><span class="kw">promoters</span>(all_genes, <span class="dt">upstream =</span> <span class="dv">0</span>, <span class="dt">downstream =</span> <span class="dv">1</span>)
all_tss =<span class="st"> </span>all_tss[<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(neg_cr<span class="op">$</span>gene)]
neg_cr[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames          ranges strand |               gene
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;
##   [1]     chr1   901460-902041      * |  ENSG00000187583.5
##   [2]     chr1 1238870-1239998      * | ENSG00000131584.14
##   -------
##   seqinfo: 17 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">all_tss[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##                      seqnames    ranges strand
##                         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    ENSG00000187583.5     chr1    901877      +
##   ENSG00000131584.14     chr1   1244989      -
##   -------
##   seqinfo: 25 sequences (1 circular) from an unspecified genome; no seqlengths</code></pre>
<p>In this example, <code>neg_cr</code> contains regions that show negative correlation between methylation and expression for the genes. The negative correlated regions are detected as:</p>
<ol style="list-style-type: decimal">
<li>extend gene to upstream 5kb and downtream 5kb;</li>
<li>for every gene, use a sliding window to go through left to right and find correlated regions by looking at the correlation between methylation in the window and expression for the gene.</li>
</ol>
<p>Since genes may be close to each other, it is possible that one correlated region for gene A overlaps with gene B, and actually we only want to overlap this correlated regions to gene A while not gene B. By specifying the mapping, we can correspond correlated regions to the correct genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_neg_cr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(neg_cr, all_tss, <span class="dt">mapping_column =</span> <span class="st">"gene"</span>, <span class="dt">w =</span> <span class="dv">50</span>, <span class="dt">mean_mode =</span> <span class="st">"w0"</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_neg_cr, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"darkgreen"</span>), <span class="dt">name =</span> <span class="st">"neg_cr"</span>, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"darkgreen"</span>))))</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-32-1.png" width="288" style="display: block; margin: auto;"></p>
<p>Similarly we can visualize the distribution of transcript to gene TSS. Since there are already connections between transcripts and their host genes, we need to provide this information when normalizing into the matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_tx =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(tx, all_tss, <span class="dt">mapping_column=</span><span class="st">"gene"</span>, <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5000</span>, <span class="dv">10000</span>), <span class="dt">w =</span> <span class="dv">50</span>, 
    <span class="dt">mean_mode =</span> <span class="st">"coverage"</span>, <span class="dt">keep =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.99</span>))
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_tx, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"black"</span>), <span class="dt">name =</span> <span class="st">"tx_coverage"</span>, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines2 =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"black"</span>))))</code></pre></div>
<p><img src="EnrichedHeatmap_files/figure-html/unnamed-chunk-33-1.png" width="288" style="display: block; margin: auto;"></p>
</div>
<div id="features-coming-from-complexheatmap-package" class="section level2">
<h2 class="hasAnchor">
<a href="#features-coming-from-complexheatmap-package" class="anchor"></a>Features coming from ComplexHeatmap package</h2>
<p>Since <strong>EnrichedHeatmap</strong> is built upon the <strong>ComplexHeatmap</strong> package, features in <strong>ComplexHeatmap</strong> can be used directly for <strong>EnrichedHeatmap</strong>. As shown before, heatmaps can be split either by <code>row_km</code> or <code>row_spilt</code> arguments.</p>
<p>The order of rows can be retrieved by <code>row_order()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
ht_list =<span class="st"> </span><span class="kw">draw</span>(ht_list)
<span class="kw">row_order</span>(ht_list)</code></pre></div>
<p>Since <code>EnrichedHeatmap</code> and <code>EnrichedHeamtapList</code> class are inherited from <code>Heamtap</code> and <code>HeamtapList</code> class respectively, all advanced parameters in the latter two classes can be directly used in the former two classes.</p>
<p>E.g. to change graphic settings for the heatmap title:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(..., <span class="dt">column_title_gp =</span> ...)</code></pre></div>
<p>To change graphic settings for legends:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(..., <span class="dt">heatmap_legend_param =</span> ...)
<span class="co"># or set is globally</span>
<span class="kw">ht_opt</span>(...)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(...)
<span class="kw">ht_opt</span>(<span class="dt">RESET =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>To set the width of the heatmaps if there are more than one heatmaps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(..., <span class="dt">width =</span> ...) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(...)</code></pre></div>
<p>For more advanced settings, please directly go to <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">the vignettes in the <strong>ComplexHeamtap</strong> package</a>.</p>
<p>Together with above features, you can make very complex heatmaps. Following example is from a real-world dataset and the details of making this plot can be found <a href="roadmap.html">in this vigentte</a>.</p>
<p>
<img src="roadmap.gif"></p>
</div>
<div id="summarize-from-a-list-of-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#summarize-from-a-list-of-matrices" class="anchor"></a>Summarize from a list of matrices</h2>
<p>Let’s assume you have a list of histone modification signals for different samples and you want to visualize the mean pattern across samples. You can first normalize histone mark signals for each sample and then calculate means values across all samples. In following example code, <code>hm_gr_list</code> is a list of <code>GRanges</code> objects which contain positions of histone modifications, <code>tss</code> is a <code>GRanges</code> object containing positions of gene TSS.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
mat_list =<span class="st"> </span><span class="ot">NULL</span>
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(hm_gr_list)) {
    mat_list[[i]] =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(hm_gr_list[[i]], tss, <span class="dt">value_column =</span> ...)
}</code></pre></div>
<p>If we compress the list of matrices as a three-dimension array where the first dimension corresponds to genes, the second dimension corresponds to windows and the third dimension corresponds to samples, the mean signal across all sample can be calculated on the third dimension. Here <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList()</a></code> simplifies this job.</p>
<p>Applying <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList()</a></code> to <code>mat_list</code>, it gives a new normalized matrix which contains mean signals across all samples and can be directly used in <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
mat_mean =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList</a></span>(mat_list)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_mean)</code></pre></div>
<p>The correlation between histone modification and gene expression can also be calculated on the third dimension of the array. In the user-defined function <code>fun</code>, <code>x</code> is the vector for gene i and window j in the array, and <code>i</code> is the index of current gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
mat_corr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList</a></span>(mat_list, <span class="dt">fun =</span> <span class="cf">function</span>(x, i) <span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(x, expr[i, ], <span class="dt">method =</span> <span class="st">"spearman"</span>))</code></pre></div>
<p>Then <code>mat_corr</code> here can be used to visualize how gene expression is correlated to histone modification around TSS.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_corr)</code></pre></div>
</div>
<div id="use-your-own-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#use-your-own-matrix" class="anchor"></a>Use your own matrix</h2>
<p><code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix()</a></code> is used to normalize the associations between genomic signals to the targets. The returned value is just a simple matrix but with several attributes attached. Sometimes, users may have their own way to generate such matrix. It is easy to add the addtional attributes and send to <code>EnrichedHeamtap()</code> for visualization.</p>
<p>Following four attributes should be attached. Basically they are used for making the axes and labels.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat, <span class="st">"upstream_index"</span>)
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat, <span class="st">"target_index"</span>)
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat, <span class="st">"downstream_index"</span>)
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat, <span class="st">"extend"</span>)</code></pre></div>
<p>To taks as an example, in following code, <code>mat2</code> is a simple matrix which only contains <code>dim</code> attributes. <code>mat2</code> can be thought as a matrix obtained from other methods.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(H3K4me3, tss, <span class="dt">value_column =</span> <span class="st">"coverage"</span>, 
    <span class="dt">extend =</span> <span class="dv">5000</span>, <span class="dt">mean_mode =</span> <span class="st">"w0"</span>, <span class="dt">w =</span> <span class="dv">50</span>)
mat2 =<span class="st"> </span>mat1
<span class="kw"><a href="https://rdrr.io/r/base/attributes.html">attributes</a></span>(mat2) =<span class="st"> </span><span class="ot">NULL</span>
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(mat2) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(mat1)
mat2[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,]    0    0    0    0
## [2,]    0    0    0    0
## [3,]    0    0    0    0
## [4,]    0    0    0    0</code></pre>
<p>As we already know, in <code>mat2</code>, upstream is extended to 5kb by 50bp window, which means the first 100 columns correspond to the upstream. Similar the last 100 columns for downstream. Here the targets is TSS which can be thought as with no width. So we can set column index attributes for upstream, target and downstream as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat2, <span class="st">"upstream_index"</span>) =<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">100</span>
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat2, <span class="st">"target_index"</span>) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/integer.html">integer</a></span>(<span class="dv">0</span>)
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat2, <span class="st">"downstream_index"</span>) =<span class="st"> </span><span class="dv">101</span><span class="op">:</span><span class="dv">200</span>
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat2, <span class="st">"extend"</span>) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5000</span>, <span class="dv">5000</span>)  <span class="co"># it must be a vector of length two</span></code></pre></div>
<p>And don’t forget to set <code>mat2</code> to <code>normalizedMatrix</code> class. And now <code>mat2</code> is a valid object for <code>EnrichedHeamtap()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(mat2) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"normalizedMatrix"</span>, <span class="st">"matrix"</span>)
mat2</code></pre></div>
<pre><code>## Normalize  to :
##   Upstream 5000 bp (100 windows)
##   Downstream 5000 bp (100 windows)
##   Not include target regions
##   720 target regions</code></pre>
<p>Above four attributes are enough for making the heatmaps, there are several more attributes which can give better information when printing <code>mat2</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat2, <span class="st">"signal_name"</span>) =<span class="st"> "H3K4me3"</span>
<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(mat2, <span class="st">"target_name"</span>) =<span class="st"> "TSS"</span>
mat2</code></pre></div>
<pre><code>## Normalize H3K4me3 to TSS:
##   Upstream 5000 bp (100 windows)
##   Downstream 5000 bp (100 windows)
##   Not include target regions
##   720 target regions</code></pre>
<p>To make the conversion easier, users can directly use <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/as.normalizedMatrix.html">as.normalizedMatrix()</a></code> for the conversion.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/attributes.html">attributes</a></span>(mat2) =<span class="st"> </span><span class="ot">NULL</span>
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(mat2) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(mat1)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/as.normalizedMatrix.html">as.normalizedMatrix</a></span>(mat2, 
    <span class="dt">k_upstream =</span> <span class="dv">100</span>, 
    <span class="dt">k_downstream =</span> <span class="dv">100</span>, 
    <span class="dt">k_target =</span> <span class="dv">0</span>,
    <span class="dt">extend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5000</span>, <span class="dv">5000</span>), 
    <span class="dt">signal_name =</span> <span class="st">"H3K4me3"</span>, 
    <span class="dt">target_name =</span> <span class="st">"TSS"</span>
)</code></pre></div>
<pre><code>## Normalize H3K4me3 to TSS:
##   Upstream 5000 bp (100 windows)
##   Downstream 5000 bp (100 windows)
##   Not include target regions
##   720 target regions</code></pre>
<p>In <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/as.normalizedMatrix.html">as.normalizedMatrix()</a></code>, you can also perform smoothing by specifying <code>smooth</code> and <code>smooth_fun</code>. Pleas check the documentation of this function.</p>
</div>
<div id="reduce-the-size-of-the-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-the-size-of-the-plot" class="anchor"></a>Reduce the size of the plot</h2>
<p>You can set <code>use_raster</code> to <code>TRUE</code> to replace the heatmap bodies with raster images. Check <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-as-raster-image" class="uri">https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-as-raster-image</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat, <span class="dt">use_raster =</span> <span class="ot">TRUE</span>, <span class="dt">raster_device =</span> ..., <span class="dt">raster_device_param =</span> ...)</code></pre></div>
</div>
<div id="error-caused-from-smoothing" class="section level2">
<h2 class="hasAnchor">
<a href="#error-caused-from-smoothing" class="anchor"></a>Error caused from smoothing</h2>
<p>If you meet following error when doing smoothing in <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix()</a></code>:</p>
<pre><code>Error: segfault from C stack overflow</code></pre>
<p>You can either:</p>
<ol style="list-style-type: decimal">
<li>remove target regions for which there are few signals overlapped to them;</li>
<li>change to another <code>smooth_fun()</code> or change parameters in <code>locfit()</code>.</li>
</ol>
<p>For solution 1, you can first calculate the matrix without smoothing and calculate the percent of <code>NA</code> values in each row. Rows having high <code>NA</code> values can be removed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># code not run</span>
mat =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(..., <span class="dt">smooth =</span> <span class="ot">FALSE</span>)
<span class="co"># the percent of NA values in each row</span>
<span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(mat, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(x)<span class="op">/</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(x)))</code></pre></div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.2
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
## 
## attached base packages:
##  [1] parallel  stats4    grid      stats     graphics  grDevices utils     datasets  methods  
## [10] base     
## 
## other attached packages:
##  [1] EnrichedHeatmap_1.15.0 GenomicRanges_1.36.1   GenomeInfoDb_1.20.0    IRanges_2.18.2        
##  [5] S4Vectors_0.22.1       BiocGenerics_0.30.0    ComplexHeatmap_2.1.0   circlize_0.4.8        
##  [9] knitr_1.25             markdown_1.1          
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.2             XVector_0.24.0         compiler_3.6.1         RColorBrewer_1.1-2    
##  [5] zlibbioc_1.30.0        bitops_1.0-6           tools_3.6.1            digest_0.6.21         
##  [9] lattice_0.20-38        evaluate_0.14          memoise_1.1.0          clue_0.3-57           
## [13] png_0.1-7              rlang_0.4.0            pkgdown_1.4.1          xfun_0.9              
## [17] GenomeInfoDbData_1.2.1 stringr_1.4.0          cluster_2.1.0          desc_1.2.0            
## [21] GlobalOptions_0.1.0    fs_1.3.1               locfit_1.5-9.1         rprojroot_1.3-2       
## [25] R6_2.4.0               GetoptLong_0.1.7       rmarkdown_1.15         magrittr_1.5          
## [29] matrixStats_0.55.0     backports_1.1.4        htmltools_0.3.6        MASS_7.3-51.4         
## [33] assertthat_0.2.1       shape_1.4.4            colorspace_1.4-1       stringi_1.4.3         
## [37] RCurl_1.95-4.12        crayon_1.3.4           rjson_0.2.20</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#make-enriched-heatmaps">Make Enriched Heatmaps</a><ul class="nav nav-pills nav-stacked">
<li><a href="#basic-usages">Basic usages</a></li>
      <li><a href="#multiple-heatmaps">Multiple heatmaps</a></li>
      <li><a href="#visualize-positive-signals-and-negative-signals-separatedly">Visualize positive signals and negative signals separatedly</a></li>
      <li><a href="#restrict-overlapping-by-mappings">Restrict overlapping by mappings</a></li>
      <li><a href="#features-coming-from-complexheatmap-package">Features coming from ComplexHeatmap package</a></li>
      <li><a href="#summarize-from-a-list-of-matrices">Summarize from a list of matrices</a></li>
      <li><a href="#use-your-own-matrix">Use your own matrix</a></li>
      <li><a href="#reduce-the-size-of-the-plot">Reduce the size of the plot</a></li>
      <li><a href="#error-caused-from-smoothing">Error caused from smoothing</a></li>
      <li><a href="#session-info">Session info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
