<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • EnrichedHeatmap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EnrichedHeatmap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/EnrichedHeatmap.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/roadmap.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../articles/row_odering.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../articles/visualize_categorical_signals_wrapper.html">UNKNOWN TITLE</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/EnrichedHeatmap">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/EnrichedHeatmap/blob/master/vignettes/roadmap.Rmd"><code>vignettes/roadmap.Rmd</code></a></small>
      <div class="hidden name"><code>roadmap.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{4. Visualize Comprehensive Associations in Roadmap dataset}
-->
<div id="visualize-comprehensive-associations-in-roadmap-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#visualize-comprehensive-associations-in-roadmap-dataset" class="anchor"></a>Visualize Comprehensive Associations in Roadmap dataset</h1>
<p><strong>Author</strong>: Zuguang Gu ( <a href="mailto:z.gu@dkfz.de">z.gu@dkfz.de</a> )</p>
<p><strong>Date</strong>: 2019-09-29</p>
<hr>
<p>In this vignette we visualize comprehensive associations between various epigenomic signals from Roadmap dataset. Here, we describe the methods and configurations used for visualizing the complicated associations.</p>
<p>The preprocessing of Roadmap data to fit into the analysis is relatively complex and is out of the scope of this package, thus here we only briefly describe the methods we use to process the data, while provide more details for the visualization part.</p>
<p>First load R packages and pre-calculated R objects which we will explain in later sections.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(EnrichedHeatmap)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GetoptLong)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(circlize)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(RColorBrewer)

<span class="kw"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span>(<span class="st">"https://jokergoo.github.io/supplementary/EnrichedHeatmap-supplementary/roadmap_normalized_matrices.RData"</span>,
    <span class="dt">destfile =</span> <span class="st">"roadmap_normalized_matrices.RData"</span>)
<span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="st">"roadmap_normalized_matrices.RData"</span>)
<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.remove</a></span>(<span class="st">"roadmap_normalized_matrices.RData"</span>)</code></pre></div>
<div id="general-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#general-overview" class="anchor"></a>General overview</h2>
<p>Roadmap dataset (<a href="http://egg2.wustl.edu/roadmap/web_portal/" class="uri">http://egg2.wustl.edu/roadmap/web_portal/</a>) covers various human cell types and tissues and has been uniformly processed. The dataset provides whole genome bisulfite sequencing data for DNA methylation, RNA sequencing data for gene expression and ChIP sequencing data for various histone modifications.</p>
<p>As we observed, gene expression shows stronger correlation pattern with DNA methylation compared to histone modifications, thus, the whole integrative analysis is centered by gene expression and methylation, with associating other histone modification signals, or in another words, we first look for regions where gene expression and methylation is correlated, and as a second step we check how histone modification signals correlate or anti-correlate to these correlated regions.</p>
<p>For the Roadmap dataset, we only use 27 samples which have both matched expression and methylation data with high data quality and have consistent subgrouping in both expression and methylation datasets (results from an unpublished analysis). Here <code>SAMPLE</code> contains annotations for samples under use and <code>COLOR</code> contains the corresponding colors for annotations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SAMPLE</code></pre></div>
<pre><code>##        id        group    sample_type  subgroup
## E016 E016          ESC PrimaryCulture subgroup1
## E003 E003          ESC PrimaryCulture subgroup1
## E024 E024          ESC PrimaryCulture subgroup1
## E007 E007     ES-deriv     ESCDerived subgroup1
## E013 E013     ES-deriv     ESCDerived subgroup1
## E012 E012     ES-deriv     ESCDerived subgroup1
## E011 E011     ES-deriv     ESCDerived subgroup1
## E004 E004     ES-deriv     ESCDerived subgroup1
## E005 E005     ES-deriv     ESCDerived subgroup1
## E006 E006     ES-deriv     ESCDerived subgroup1
## E050 E050 HSC_&amp;_B-cell    PrimaryCell subgroup2
## E112 E112       Thymus  PrimaryTissue subgroup2
## E071 E071        Brain  PrimaryTissue subgroup2
## E100 E100       Muscle  PrimaryTissue subgroup2
## E104 E104        Heart  PrimaryTissue subgroup2
## E095 E095        Heart  PrimaryTissue subgroup2
## E105 E105        Heart  PrimaryTissue subgroup2
## E065 E065        Heart  PrimaryTissue subgroup2
## E109 E109    Digestive  PrimaryTissue subgroup2
## E106 E106    Digestive  PrimaryTissue subgroup2
## E079 E079    Digestive  PrimaryTissue subgroup2
## E094 E094    Digestive  PrimaryTissue subgroup2
## E097 E097        Other  PrimaryTissue subgroup2
## E066 E066        Other  PrimaryTissue subgroup2
## E098 E098        Other  PrimaryTissue subgroup2
## E096 E096        Other  PrimaryTissue subgroup2
## E113 E113        Other  PrimaryTissue subgroup2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">COLOR</code></pre></div>
<pre><code>## $group
##          ESC     ES-deriv HSC_&amp;_B-cell       Thymus        Brain       Muscle        Heart 
##    "#924965"    "#4178AE"    "#678C69"    "#DAB92E"    "#C5912B"    "#C2655D"    "#D56F80" 
##    Digestive        Other 
##    "#C58DAA"    "#999999" 
## 
## $sample_type
## PrimaryCulture     ESCDerived    PrimaryCell  PrimaryTissue 
##      "#8cd2c7"      "#bfb9db"      "#faf7b4"      "#f57f73" 
## 
## $subgroup
## subgroup1 subgroup2 
## "#A6CEE3" "#1F78B4"</code></pre>
<p>The 27 samples are separated into two subgroups (labeled as <code>subgroup</code> column in <code>SAMPLE</code>) where one subgroup (<code>subgroup1</code>) corresponds to embryonic stem cells and the other subgroup (<code>subgroup2</code>) corresponds to primary tissues or mature cells. Samples in the two subgroups show distinct difference in both expression and methylation datasets (results from an unpublished analysis).</p>
<p>In this supplementary, we visualize the enrichment of various epigenomic signals around gene TSS with upstream 5kb and downstream 10kb. All epigenomic signals have already been normalized to gene TSS and stored as R objects because the calculation for the matrices involves complicated data preprocessing and normalizing all datasets to gene TSS is time comsuing. However, we will still provide pseudo code which generates these normalized matrices.</p>
<p>All epigenomic signals are normalized to gene TSS with same settings (upstream 5kb, downstream 10kb, window size 50bp), thus, all normalized matrices have the same dimension and row order, and i^th row, j^th column in all matrices correspond to a same position relative to a same gene.</p>
<p>There are following pre-calculated normalized matrices. Here we only simply describe these objects and details for generating them will be explained in later sections.</p>
<ul>
<li>
<code>mat_neg_cr</code>: a normalized matrix for correlated regions (CRs, The definition of CR is introduced in following sections) showing significant negative correlation between methylation and gene expression. The value in the matrix is whether a window is covered by negative CRs (values are 0 or 1).</li>
<li>
<code>meth_mat_corr</code>: a normalized matrix for all CRs. The value in the matrix is the mean correlation for CRs overlapped to a window.</li>
<li>
<code>meth_mat_mean</code>: a normalized matrix for mean methylation across all samples.</li>
<li>
<code>meth_mat_diff</code>: a normalized matrix for mean methylation difference between two subgroups.</li>
<li>
<code>mat_cgi</code>: a normalized matrix for CpG islands. The value in the matrix is whether a window is covered by CGIs.</li>
<li>
<code>hist_mat_corr_list</code>: a list of normalized matrices for the correlation between histone modification signals and gene expression. Each matrix corresponds to one type of histone modification.</li>
<li>
<code>hist_mat_mean_list</code>: a list of normalized matrices for the mean histone modification signals across all samples.</li>
<li>
<code>hist_mat_diff_list</code>: a list of normalized matrices for the histone modification signal difference between two subgroups.</li>
</ul>
<p>Other R objects are</p>
<ul>
<li>
<code>gene</code>: A <code>GRanges</code> object which contains positions of genes.</li>
<li>
<code>tss</code>: A <code>GRanges</code> object which contains positions of gene TSS.</li>
<li>
<code>gene_symbol</code> a mapping between Ensembl IDs and gene symbols.</li>
<li>
<code>expr</code>: gene expression matrix (Values are measured by <code><a href="https://rdrr.io/r/base/Log.html">log2(RPKM + 1)</a></code>).</li>
</ul>
<p><code>tss</code> and <code>expr</code> also have the same row order as normalized matrices.</p>
</div>
<div id="association-between-gene-expression-and-methylation" class="section level2">
<h2 class="hasAnchor">
<a href="#association-between-gene-expression-and-methylation" class="anchor"></a>Association between gene expression and methylation</h2>
<p>When looking for associations between DNA methylation and gene expression, the process is gene centric. We briefly describe the method as follows:</p>
<p>Each gene is extended to upstream 50kb and downstream 50kb to the full gene body. For each extended gene, we use a 6-CpG sliding window with step of 3 CpGs and with maximum window size of 10kb. In each window, mean methylation is calculated from the 6-CpG sites and the Spearman correlation as well as the correlaion test to expression of the current gene is calculated. We term these 6 CpG windows as correlated regions (CRs) and significant CRs are filterred by FDR &lt; 0.05 (from the correlation test) and methylation difference between the two subgroups are larger than 0.2.</p>
<p>According to this procedure, each CR belongs to one certain gene, which means, there is a mapping between CRs and genes. Thus when normalizing CRs to TSS, this mapping should be provided so that CRs can be correctly normalized to their host genes. Or else there can be scenarios that two extended genes overlap to each other and one CR of gene A also overlaps to gene B. If the mapping is not provided, this CR will be wrongly mapped to gene B.</p>
<p>Assume <code>cr</code> contains CRs for all 6-CpG windows and the gene name column is named as <code>gene_id</code>, the correlation column is named as <code>corr</code>, following code normalizes CRs to TSS.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
mat_corr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(cr, tss, <span class="dt">mapping_column =</span> <span class="st">"gene_id"</span>, <span class="dt">value_column =</span> <span class="st">"corr"</span>, 
    <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, ...)</code></pre></div>
<p>Assume <code>sig_neg_cr</code> contains CRs showing significant negative correlations to expression, following code normalizes significant negative CRs to TSS. Note there is no <code>value_column</code> in following code so that the normalized matrix measures whether each window is covered by <code>sig_neg_cr</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
mat_neg_cr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(sig_neg_cr, tss, <span class="dt">mapping_column =</span> <span class="st">"gene_id"</span>, 
    <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, ...)</code></pre></div>
<p>The correlation itself does not tell the methylation level (highly methylated or lowly methylated) nor the variability of methylation among samples. In order to get a more comprehensive view of the methylation, we also normalize mean methylation and methylation variability to TSS. Methylation data represents as a matrix where rows are CpG sites and columns are samples, thus we directly calculate mean methylation among samples with row means, then normalize the mean methylation to gene TSS.</p>
<p>In following code, assume <code>meth</code> is a <code>GRanges</code> object of CpG sites and meta data columns contain methylation matrix for all samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
meth_mean =<span class="st"> </span>meth
<span class="kw">mcols</span>(meth_mean) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">mean_meth =</span> <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="kw">mcols</span>(meth)))
meth_mat_mean =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth_mean, tss, <span class="dt">value_column =</span> <span class="st">"mean_meth"</span>, 
    <span class="dt">mode =</span> <span class="st">"absolute"</span>, ...)</code></pre></div>
<p>And the mean methylation difference is calculated as <span class="math inline">\(m_1 - m_2\)</span> where <span class="math inline">\(m_1\)</span> is the mean methylation matrix in subgroup 1 and <span class="math inline">\(m_2\)</span> is the mean methylation in subgroup 2.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
meth_mean_<span class="dv">1</span> =<span class="st"> </span>meth
<span class="kw">mcols</span>(meth_mean_<span class="dv">1</span>) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">mean_meth =</span> <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="kw">mcols</span>(meth[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup1"</span>])))
meth_mat_mean_<span class="dv">1</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth_mean_<span class="dv">1</span>, tss, <span class="dt">value_column =</span> <span class="st">"mean_meth"</span>, 
    <span class="dt">mode =</span> <span class="st">"absolute"</span>, ...)
meth_mean_<span class="dv">2</span> =<span class="st"> </span>meth
<span class="kw">mcols</span>(meth_mean_<span class="dv">2</span>) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">mean_meth =</span> <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="kw">mcols</span>(meth[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup2"</span>])))
meth_mat_mean_<span class="dv">2</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth_mean_<span class="dv">2</span>, tss, <span class="dt">value_column =</span> <span class="st">"mean_meth"</span>, 
    <span class="dt">mode =</span> <span class="st">"absolute"</span>, ...)
meth_mat_diff =<span class="st"> </span>meth_mat_mean_<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>meth_mat_mean_<span class="dv">2</span></code></pre></div>
<p>Since CRs are detected in upstream 50kb and downstream 50kb of the full gene, while for the visualizaiton, only upstream 5kb and downstream 10kb of TSS are used, genes which do not have a significant negative CR in [-5kb, 10kb] of TSS are removed (can be filtered by <code>rowSums(mat_neg_cr) &gt; 0</code>), which finnally results in 1832 genes for the analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(tss)</code></pre></div>
<pre><code>## [1] 1832</code></pre>
</div>
<div id="association-with-histone-modifications" class="section level2">
<h2 class="hasAnchor">
<a href="#association-with-histone-modifications" class="anchor"></a>Association with histone modifications</h2>
<p>We use following four types of histone modifications which show specific patterns at gene TSS: H3K4me1, H3K4me3, H3K27ac and H3K27me3. Similar as methylation, for each type of histone modification, there are three matrices which are 1. correlation to gene expression; 2. mean signals across all samples; and 3. the mean signal difference between two subgroups.</p>
<p>Being different from methylation datasets, the peak regions of histone modification data across all samples cannot locate at a same genomic position which makes it naturelly not a matrix-like data. To format it, we normalize signals to gene TSS for eash sample separately, which generates a list of matrices with same dimensions and settings.</p>
<p>Assume <code>peak</code> is a list of <code>GRanges</code> objects for peak regions in different samples. In following, we additionally set <code>keep = c(0, 0.99)</code> to adjust outlier values which are larger than 99th percentile.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
<span class="cf">for</span>(i <span class="cf">in</span> n_sample) {
    <span class="co"># assume the column name for the signals is called 'density'</span>
    hm_list[[i]] =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(peak[[i]], tss, <span class="dt">value_column =</span> <span class="st">"density"</span>, <span class="dt">keep =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.99</span>))
}</code></pre></div>
<p>If we compress the list of matrices as a three-dimension array where the first dimension corresponds to genes, the second dimension corresponds to windows and the third dimension corresponds to samples, the mean signal across all sample can be calculated on the third dimension. Here <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList()</a></code> simplifies this job.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
hm_mat_mean =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList</a></span>(hm_list, mean)</code></pre></div>
<p>The mean difference between two subgroups can be calculated in a similar way:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
<span class="co"># hm_list_1 and hm_list_2 are normalized matrices for subgroup1 and subgroup2 separatedly</span>
hm_mat_mean_<span class="dv">1</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList</a></span>(hm_list_<span class="dv">1</span>, mean)
hm_mat_mean_<span class="dv">2</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList</a></span>(hm_list_<span class="dv">2</span>, mean)
hm_mat_diff =<span class="st"> </span>hm_mat_mean_<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>hm_mat_mean_<span class="dv">2</span></code></pre></div>
<p>The correlation between histone modification and gene expression can also be calculated on the third dimension of the array. In the user-defined function <code>fun</code>, <code>x</code> is the vector for gene i and window j in the array, and <code>i</code> is the index of current gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
hm_corr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/getSignalsFromList.html">getSignalsFromList</a></span>(hm_list, <span class="dt">fun =</span> <span class="cf">function</span>(x, i) {
    <span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(x, expr[i, ], <span class="dt">method =</span> <span class="st">"spearman"</span>) <span class="co"># x = array[i, j, ]</span>
})</code></pre></div>
<p>We apply this method on all four types of histone modifications and normalized matrices are stored as <code>hist_mat_corr_list</code>, <code>hist_mat_mean_list</code> and <code>hist_mat_diff_list</code>. All three objects are list of four matrices.</p>
</div>
<div id="normalize-cgi-to-gene-tss" class="section level2">
<h2 class="hasAnchor">
<a href="#normalize-cgi-to-gene-tss" class="anchor"></a>normalize CGI to gene TSS</h2>
<p>Normalizing CpG islands to TSS is straightforward. The value in the normalized matrix is whether each window is covered by CpG islands.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this chunk of code is only for demonstration</span>
mat_cgi =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(cgi, tss, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>, ...)</code></pre></div>
</div>
<div id="order-and-group-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#order-and-group-genes" class="anchor"></a>Order and group genes</h2>
<p>The ordering of genes and sometimes separating genes into several groups are important to strenghen the effect of visualization of the patterns. Especially when we have multiple heatmaps which share same row order, a proper way to order genes is more important to highlight patterns for all heatmaps.</p>
<p>Which method to order genes depends on what pattern users want to reveal. In this analysis, the message we want to show from the heatmaps are 1. the enrichment of significantly negatively correlated regions (negCRs) around TSS; 2. the difference between subgroup 1 and subgroup 2; and 3. the different methylation patterns for different genes. The following procedures shows how we group genes and how we order genes for better showing these patterns.</p>
<p>The expression difference between subgroup 1 and subgroup 2 samples is a major grouping factor. We construct a category vector which corresponds to high expression and low expression in subgroup 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expr_mean =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(expr[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup1"</span>]) <span class="op">-</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(expr[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup2"</span>])
expr_split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(expr_mean <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">"high"</span>, <span class="st">"low"</span>)
expr_split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(expr_split, <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"high"</span>, <span class="st">"low"</span>))</code></pre></div>
<p>After looking at the methylation data, we found the methylation shows big difference between genes and the major difference happens at small areas around TSS and flanks more to the downstream of TSS. Thus, we only extract 20% of upstream of TSS and 40% of downstream of TSS and use methylation to separate genes into two subgroups which are low TSS-methylation group and high TSS-methylation group.</p>
<p>The label of methylation groups are adjsuted that the first group always has lowest mean TSS-methylation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">123</span>)
upstream_index =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/attr.html">attr</a></span>(meth_mat_mean, <span class="st">"upstream_index"</span>))
meth_split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(meth_mat_mean[, <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(upstream_index<span class="op">*</span><span class="fl">0.8</span>), <span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(upstream_index<span class="op">*</span><span class="fl">1.4</span>))], 
    <span class="dt">centers =</span> <span class="dv">2</span>)<span class="op">$</span>cluster
x =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(meth_mat_mean[, <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(upstream_index<span class="op">*</span><span class="fl">0.8</span>), <span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(upstream_index<span class="op">*</span><span class="fl">1.4</span>))]), 
    meth_split, mean)
od =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/structure.html">structure</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(x), <span class="dt">names =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(x))
meth_split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"cluster"</span>, od[<span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(meth_split)])</code></pre></div>
<p>Both grouping from expression and methylation is important, thus To make a more informative grouping of genes, the split from expression and methylation are combined:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">combined_split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(meth_split, expr_split, <span class="dt">sep =</span> <span class="st">"|"</span>)</code></pre></div>
<p>There is one combined group with too few number of genes</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tb =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(combined_split)
tb</code></pre></div>
<pre><code>## combined_split
## cluster1|high  cluster1|low cluster2|high  cluster2|low 
##           306           940            25           561</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tb[<span class="st">"cluster2|high"</span>]<span class="op">/</span><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(tb)</code></pre></div>
<pre><code>## cluster2|high 
##    0.01364629</code></pre>
<p>The proprotion of “cluster2|high” group is too small and also we don’t want to make too many row clusters, in order to make the plot more clear, “cluster2|high” is removed from the analysis.</p>
<p>Also all related variables should be subsetted to remove “cluster2|high” genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">l =<span class="st"> </span>combined_split <span class="op">!=</span><span class="st"> "cluster2|high"</span>
tss =<span class="st"> </span>tss[l]
expr =<span class="st"> </span>expr[l, ]
hist_mat_corr_list =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(hist_mat_corr_list, <span class="cf">function</span>(x) x[l, ])
hist_mat_mean_list =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(hist_mat_mean_list, <span class="cf">function</span>(x) x[l, ])
hist_mat_diff_list =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(hist_mat_diff_list, <span class="cf">function</span>(x) x[l, ])
mat_neg_cr =<span class="st"> </span>mat_neg_cr[l, ]
mat_cgi =<span class="st"> </span>mat_cgi[l, ]
meth_mat_corr =<span class="st"> </span>meth_mat_corr[l, ]
meth_mat_mean =<span class="st"> </span>meth_mat_mean[l, ]
meth_mat_diff =<span class="st"> </span>meth_mat_diff[l, ]
expr_split =<span class="st"> </span>expr_split[l]
meth_split =<span class="st"> </span>meth_split[l]
combined_split =<span class="st"> </span>combined_split[l]
n_row_cluster =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(combined_split))</code></pre></div>
<p>Next we calculate row order, following is a way which shows clear patterns for all signals.</p>
<p>From heatmaps which we show later, there is a clear pattern that the negCRs are enriched at consistent positions downstream of TSS. To show this specific pattern, we designed a specific distance metric which calculates how close the negCRs on two genes are based on relative distance to TSS.</p>
<p>For two rows in the normalized matrix, assume <span class="math inline">\(a_1, a_2, ..., a_{n_1}\)</span> are the window indices for one gene which overlaps with negative correlated regions and <span class="math inline">\(b_1, b_2, ... b_{n_2}\)</span> are the indices for the other gene, the distance which is based on closeness of the overlapped windows in the two genes is defined as:</p>
<p><span class="math display">\[ d_{closeness} = \frac{\sum_{i=1}^{n_1} \sum_{j=1}^{n_2} {|a_i - b_j|}}{n_1 \cdot n_2}\]</span></p>
<p>Following code calculates row orders for genes. For each row cluster split by <code>combined_split</code>, rows are clustered separately. In following code, <code>l_list</code> is a logical partition of genes and for each row cluster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">merge_row_order =<span class="st"> </span><span class="cf">function</span>(l_list) {
    <span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"c"</span>, <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(l_list, <span class="cf">function</span>(l) {
        <span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(l) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/integer.html">integer</a></span>(<span class="dv">0</span>))
        <span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(l) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(l))
        dend1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/dist_by_closeness.html">dist_by_closeness</a></span>(mat_neg_cr[l, ])))
        dend1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span>(dend1, <span class="op">-</span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/enriched_score.html">enriched_score</a></span>(mat_neg_cr[l, ]))
        od =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/order.dendrogram.html">order.dendrogram</a></span>(dend1)
        <span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(l)[od]
    }))
}

row_order =<span class="st"> </span><span class="kw">merge_row_order</span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
    combined_split <span class="op">==</span><span class="st"> "cluster1|high"</span>,
    combined_split <span class="op">==</span><span class="st"> "cluster1|low"</span>,
    combined_split <span class="op">==</span><span class="st"> "cluster2|low"</span>
))</code></pre></div>
</div>
<div id="organize-heatmaps" class="section level2">
<h2 class="hasAnchor">
<a href="#organize-heatmaps" class="anchor"></a>Organize heatmaps</h2>
<p>After all matrics are generated, we can generate the complex heatmap list.</p>
<p>First we prepare the heatmap for expression. The columns of expression matrix are only clustered for each subgroup.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dend1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(expr[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup1"</span>]))))
hc1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/as.hclust.html">as.hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span>(dend1, <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span>(expr[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup1"</span>])))
expr_col_od1 =<span class="st"> </span>hc1<span class="op">$</span>order
dend2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(expr[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup2"</span>]))))
hc2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/as.hclust.html">as.hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span>(dend2, <span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span>(expr[, SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup2"</span>])))
expr_col_od2 =<span class="st"> </span>hc2<span class="op">$</span>order
expr_col_od =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup1"</span>)[expr_col_od1], 
              <span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(SAMPLE<span class="op">$</span>subgroup <span class="op">==</span><span class="st"> "subgroup2"</span>)[expr_col_od2])</code></pre></div>
<p>The first heatmap is the expression which is a normal heatmap where we put sample annotations on top.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(expr, <span class="dt">name =</span> <span class="st">"expr"</span>, <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>,
    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">4</span>, <span class="st">"cm"</span>), <span class="dt">show_column_dend =</span> <span class="ot">FALSE</span>, 
    <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">column_order =</span> expr_col_od,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">df =</span> SAMPLE[, <span class="op">-</span><span class="dv">1</span>], <span class="dt">col =</span> COLOR, 
        <span class="dt">annotation_name_side =</span> <span class="st">"left"</span>),
    <span class="dt">column_title =</span> <span class="st">"Expression"</span>, <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>),
    <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>, <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>We extract top 20 genes with most significant p-values simply by t-test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(genefilter)
df =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/genefilter/man/rowFtests.html">rowttests</a></span>(expr, <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(SAMPLE<span class="op">$</span>subgroup))
top_genes =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(df[<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(df<span class="op">$</span>p.value)[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>], ])</code></pre></div>
<p>These top genes are added as a text annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index =<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(expr) <span class="op">%in%</span><span class="st"> </span>top_genes)
labels =<span class="st"> </span>gene_symbol[<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(expr)[index]]
ht_list =<span class="st"> </span><span class="kw">rowAnnotation</span>(<span class="dt">sig_gene =</span> <span class="kw">anno_mark</span>(<span class="dt">at =</span> index, <span class="dt">labels =</span> labels,
        <span class="dt">side =</span> <span class="st">"left"</span>, <span class="dt">labels_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">10</span>), <span class="dt">padding =</span> <span class="fl">0.5</span>, 
        <span class="dt">extend =</span> <span class="kw">unit</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">0</span>), <span class="st">"cm"</span>))) <span class="op">+</span><span class="st"> </span>ht_list</code></pre></div>
<p>On the right side of the expression is a row annotation which show the length of genes, constructed by <code>rowAnnotation()</code> function and appended to the heatmap list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gl =<span class="st"> </span><span class="kw">width</span>(gene[<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(tss)])
gl[gl <span class="op">&gt;</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(gl, <span class="fl">0.95</span>)] =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(gl, <span class="fl">0.95</span>)
ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw">rowAnnotation</span>(<span class="dt">gene_len =</span> <span class="kw">anno_points</span>(gl, <span class="dt">size =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">"mm"</span>), 
        <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"#00000040"</span>), 
        <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">1e5</span>, <span class="fl">2e5</span>), <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"0bp"</span>, <span class="st">"100bp"</span>, <span class="st">"200bp"</span>)), 
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="fl">1.5</span>, <span class="st">"cm"</span>)))</code></pre></div>
<p>Add the heatmap which shows enrichment of CGI to TSS. Top annotation which shows enrichment pattern is added by <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched()</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">axis_name =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"-5kb"</span>, <span class="st">"TSS"</span>, <span class="st">"10kb"</span>)
ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_cgi, <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"darkorange"</span>), <span class="dt">name =</span> <span class="st">"CGI"</span>,
    <span class="dt">column_title =</span> <span class="st">"CGI"</span>, <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>),
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"darkorange"</span>, 
        <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))), 
    <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>) </code></pre></div>
<p>Add the heatmap which shows the correlation between methylation and expression. Since in the normalized matrix <code>meth_mat_corr</code>, there are positive correlations and negative correlations, both <code>pos_col</code> and <code>neg_col</code> are set so that enrichment for positive correlation and negative correlation are drawn separately in the top annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bg_col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="dv">8</span>, <span class="st">"Set2"</span>)
cor_col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"darkgreen"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(meth_mat_corr, <span class="dt">col =</span> cor_col_fun, <span class="dt">name =</span> <span class="st">"meth_corr"</span>, 
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">pos_col =</span> <span class="st">"red"</span>, 
            <span class="dt">neg_col =</span> <span class="st">"darkgreen"</span>, <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), 
        <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))), 
    <span class="dt">column_title =</span> <span class="st">"meth_corr"</span>, <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fill =</span> bg_col[<span class="dv">1</span>]),
        <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Add the heatmap which shows mean methylation among all samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">meth_col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(meth_mat_mean, <span class="dt">col =</span> meth_col_fun, <span class="dt">name =</span> <span class="st">"meth_mean"</span>, 
    <span class="dt">column_title =</span> <span class="st">"meth_mean"</span>, <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fill =</span> bg_col[<span class="dv">1</span>]),
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"red"</span>, 
        <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))),
    <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>For the heatmap showing difference between subgroups, we define a function which generate color mappings showing symmetric color mapping for positive difference and negative difference. This works for both methylation difference and histone modification difference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">generate_diff_color_fun =<span class="st"> </span><span class="cf">function</span>(x) {
    q =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(x, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.95</span>))
    max_q =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(q))
    <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span>max_q, <span class="dv">0</span>, max_q), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#3794bf"</span>, <span class="st">"#FFFFFF"</span>, <span class="st">"#df8640"</span>))
}

ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(meth_mat_diff, <span class="dt">name =</span> <span class="st">"meth_diff"</span>, 
    <span class="dt">col =</span> <span class="kw">generate_diff_color_fun</span>(meth_mat_diff),
    <span class="dt">column_title =</span> <span class="st">"meth_diff"</span>, <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fill =</span> bg_col[<span class="dv">1</span>]),
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">pos_col =</span> <span class="st">"#df8640"</span>, 
            <span class="dt">neg_col =</span> <span class="st">"#3794bf"</span>, <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), 
        <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))),
    <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Since here in the final heamtap list, there are 17 heatmaps which are too many to put in one row layout. Thus, we separate the heatmaps into two and assign them to <code>ht_list_1</code> and <code>ht_list_2</code>. In following code, the heatmaps for the last three histone modifications are assigned to the second heatmap list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list_<span class="dv">2</span> =<span class="st"> </span><span class="ot">NULL</span>
ht_list_<span class="dv">1</span> =<span class="st"> </span><span class="ot">NULL</span>
mark_name =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(hist_mat_corr_list)
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(hist_mat_corr_list)) {
    <span class="co"># heatmaps for the 2nd, 3th and 4th histone modifications are assigned to a new `ht_list`</span>
    <span class="cf">if</span>(i <span class="op">==</span><span class="st"> </span><span class="dv">2</span>) {
        ht_list_<span class="dv">1</span> =<span class="st"> </span>ht_list
        ht_list =<span class="st"> </span><span class="ot">NULL</span>
    }

    ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(hist_mat_corr_list[[i]], <span class="dt">col =</span> cor_col_fun, 
        <span class="dt">name =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"@{mark_name[i]}_corr"</span>), <span class="dt">column_title =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"@{mark_name[i]}_corr"</span>), 
        <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fill =</span> bg_col[i<span class="op">+</span><span class="dv">1</span>]),
        <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">pos_col =</span> <span class="st">"red"</span>,
                <span class="dt">neg_col =</span> <span class="st">"darkgreen"</span>, <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), 
            <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))), 
        <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)

    ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(hist_mat_mean_list[[i]], 
        <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="kw"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(hist_mat_mean_list[[i]], <span class="fl">0.95</span>)), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"purple"</span>)), 
        <span class="dt">name =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"@{mark_name[i]}_mean"</span>), <span class="dt">column_title =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"@{mark_name[i]}_mean"</span>), 
        <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fill =</span> bg_col[i<span class="op">+</span><span class="dv">1</span>]),
        <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">"purple"</span>, 
            <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))),
        <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)

    ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(hist_mat_diff_list[[i]], 
        <span class="dt">col =</span> <span class="kw">generate_diff_color_fun</span>(hist_mat_diff_list[[i]]), 
        <span class="dt">name =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"@{mark_name[i]}_diff"</span>), <span class="dt">column_title =</span> <span class="kw"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html">qq</a></span>(<span class="st">"@{mark_name[i]}_diff"</span>), 
        <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fill =</span> bg_col[i<span class="op">+</span><span class="dv">1</span>]), 
        <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">lines =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">pos_col =</span> <span class="st">"#df8640"</span>, 
                <span class="dt">neg_col =</span> <span class="st">"#3794bf"</span>, <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span>n_row_cluster), 
            <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>, <span class="dt">facing =</span> <span class="st">"inside"</span>))),
        <span class="dt">axis_name =</span> axis_name, <span class="dt">axis_name_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), <span class="dt">use_raster =</span> <span class="ot">TRUE</span>)
}
ht_list_<span class="dv">2</span> =<span class="st"> </span>ht_list</code></pre></div>
<p>We assign same <code>split</code> and <code>row_order</code> to both heatmap lists so that they can be correctedly corresponded.</p>
<p>All heatmaps are split by <code>combined_split</code>. Here we rename values in <code>combined_split</code> to <code>cluster1</code>, <code>cluster2</code> and <code>cluster3</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span>(combined_split)
split[combined_split <span class="op">==</span><span class="st"> "cluster1|high"</span>] =<span class="st"> "cluster1"</span>
split[combined_split <span class="op">==</span><span class="st"> "cluster1|low"</span>] =<span class="st"> "cluster2"</span>
split[combined_split <span class="op">==</span><span class="st"> "cluster2|low"</span>] =<span class="st"> "cluster3"</span></code></pre></div>
<p>For each heatmap list, a single column heatmap (or bar) is attached to the most left side to represent difference of expression between subgroups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list_<span class="dv">1</span> =<span class="st"> </span><span class="kw">Heatmap</span>(expr_split, <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">name =</span> <span class="st">"expr_diff"</span>, 
  <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"high"</span> =<span class="st"> "red"</span>, <span class="st">"low"</span> =<span class="st"> "darkgreen"</span>), 
  <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">"mm"</span>)) <span class="op">+</span><span class="st"> </span>ht_list_<span class="dv">1</span></code></pre></div>
<p>Now we explictly use <code>draw()</code> function to make the heatmaps because there are some global settings for all heatmaps. We also decorate the final heatmap such as adding labels and axes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list_<span class="dv">1</span> =<span class="st"> </span><span class="kw">draw</span>(ht_list_<span class="dv">1</span>, 
    <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">row_order =</span> row_order, <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>,
    <span class="dt">row_split =</span> split, <span class="dt">heatmap_legend_side =</span> <span class="st">"bottom"</span>, <span class="dt">ht_gap =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">"mm"</span>))

add_boxplot_of_gene_length =<span class="st"> </span><span class="cf">function</span>(ht_list) {
  
    row_order_list =<span class="st"> </span><span class="kw">row_order</span>(ht_list)
    lt =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(row_order_list, <span class="cf">function</span>(ind) gl[ind])
    bx =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(lt, <span class="dt">plot =</span> <span class="ot">FALSE</span>)<span class="op">$</span>stats
    n =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(row_order_list)
    
    <span class="kw">decorate_annotation</span>(<span class="st">"gene_len"</span>, <span class="dt">slice =</span> <span class="dv">1</span>, {
        rg =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/range.html">range</a></span>(bx)
        rg[<span class="dv">1</span>] =<span class="st"> </span>rg[<span class="dv">1</span>] <span class="op">-</span><span class="st"> </span>(rg[<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>rg[<span class="dv">1</span>])<span class="op">*</span><span class="fl">0.1</span>
        rg[<span class="dv">2</span>] =<span class="st"> </span>rg[<span class="dv">2</span>] <span class="op">+</span><span class="st"> </span>(rg[<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>rg[<span class="dv">1</span>])<span class="op">*</span><span class="fl">0.1</span>
        <span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">y =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">"npc"</span>) <span class="op">+</span><span class="st"> </span><span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">"mm"</span>), <span class="dt">just =</span> <span class="st">"bottom"</span>, 
            <span class="dt">height =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">"cm"</span>), <span class="dt">yscale =</span> rg, <span class="dt">xscale =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.5</span>, n <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>)))
        <span class="kw">grid.rect</span>()
        <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) {
            <span class="kw">grid.boxplot</span>(<span class="dt">pos =</span> i, lt[[i]], <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> i), <span class="dt">outline =</span> <span class="ot">FALSE</span>)
        }
        <span class="kw">grid.text</span>(<span class="st">"Gene length"</span>, <span class="dt">y =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">"npc"</span>) <span class="op">+</span><span class="st"> </span><span class="kw">unit</span>(<span class="fl">2.5</span>, <span class="st">"mm"</span>), 
            <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>), <span class="dt">just =</span> <span class="st">"bottom"</span>)
        <span class="kw">upViewport</span>() 
    })
}

<span class="kw">add_boxplot_of_gene_length</span>(ht_list_<span class="dv">1</span>)</code></pre></div>
<p><img src="roadmap_files/figure-html/unnamed-chunk-27-1.png" width="1344"></p>
<p><code>split</code> and <code>row_order</code> are set to the second heatmap list as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list_<span class="dv">2</span> =<span class="st"> </span><span class="kw">Heatmap</span>(expr_split, <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">name =</span> <span class="st">"expr_diff"</span>, 
    <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"high"</span> =<span class="st"> "red"</span>, <span class="st">"low"</span> =<span class="st"> "darkgreen"</span>), 
    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">"mm"</span>)) <span class="op">+</span><span class="st"> </span>ht_list_<span class="dv">2</span>
ht_list_<span class="dv">2</span> =<span class="st"> </span><span class="kw">draw</span>(ht_list_<span class="dv">2</span>,
    <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">row_order =</span> row_order, <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>,
    <span class="dt">row_split =</span> split, <span class="dt">heatmap_legend_side =</span> <span class="st">"bottom"</span>, <span class="dt">ht_gap =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">"mm"</span>))</code></pre></div>
<p><img src="roadmap_files/figure-html/unnamed-chunk-28-1.png" width="1344"></p>
<p>Note rows in the two heatmap list are all the same.</p>
<p>Following code only extracts the annotation graphics. We put all the annotation graphics in a layout so that it is easily to compare between them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">add_anno_enriched =<span class="st"> </span><span class="cf">function</span>(ht_list, name, ri, ci) {
    <span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">layout.pos.row =</span> ri, <span class="dt">layout.pos.col =</span> ci))
    <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/extract_anno_enriched.html">extract_anno_enriched</a></span>(ht_list, name, <span class="dt">newpage =</span> <span class="ot">FALSE</span>)
    <span class="kw">upViewport</span>()
}

<span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">layout =</span> <span class="kw">grid.layout</span>(<span class="dt">nr =</span> <span class="dv">3</span>, <span class="dt">nc =</span> <span class="dv">6</span>)))
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"meth_corr"</span>,     <span class="dv">1</span>, <span class="dv">1</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"meth_mean"</span>,     <span class="dv">1</span>, <span class="dv">2</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"meth_diff"</span>,     <span class="dv">1</span>, <span class="dv">3</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"CGI"</span>,           <span class="dv">1</span>, <span class="dv">4</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"H3K4me3_corr"</span>,  <span class="dv">2</span>, <span class="dv">1</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"H3K4me3_mean"</span>,  <span class="dv">2</span>, <span class="dv">2</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">1</span>, <span class="st">"H3K4me3_diff"</span>,  <span class="dv">2</span>, <span class="dv">3</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K4me1_corr"</span>,  <span class="dv">2</span>, <span class="dv">4</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K4me1_mean"</span>,  <span class="dv">2</span>, <span class="dv">5</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K4me1_diff"</span>,  <span class="dv">2</span>, <span class="dv">6</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K27ac_corr"</span>,  <span class="dv">3</span>, <span class="dv">1</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K27ac_mean"</span>,  <span class="dv">3</span>, <span class="dv">2</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K27ac_diff"</span>,  <span class="dv">3</span>, <span class="dv">3</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K27me3_corr"</span>, <span class="dv">3</span>, <span class="dv">4</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K27me3_mean"</span>, <span class="dv">3</span>, <span class="dv">5</span>)
<span class="kw">add_anno_enriched</span>(ht_list_<span class="dv">2</span>, <span class="st">"H3K27me3_diff"</span>, <span class="dv">3</span>, <span class="dv">6</span>)
<span class="kw">upViewport</span>()</code></pre></div>
<p><img src="roadmap_files/figure-html/unnamed-chunk-29-1.png" width="1344"></p>
</div>
<div id="interpretation" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation" class="anchor"></a>Interpretation</h2>
<p>Following figure puts the two heatmap lists into one plotting page and arranges legends to the right of the heatmaps.</p>
<p><img src="roadmap.gif"></p>
<p>Generally, genes in cluster 1 and 2 have high expression, long gene length (annotation “Gene length”) and low methylation over TSS (heatmap “meth_mean”) which correspond well with the enrichment of CpG islands over TSS (heatmap “CGI”), while genes in cluster 3 have low expression, short gene length, and intermediate mean methylation with almost none CGIs overlapping TSS. There is enrichment for significant negative CRs (negCRs) downstream of TSS in cluster 1 and cluster 2 (solid and dashed green lines in annotation of “meth_corr” heatmap, the peaks of the enrichment locate at approximately +2kb of TSS.) while for cluster 3 genes, the enrichment of negCRs is very close to TSS. By associating the heatmap “CGI”, “meth_corr”, “meth_mean” and “meth_diff” together, we can make the conclusion that for genes in cluster 1 and cluster 2, negCRs are enriched at the downstream border of CGI over TSS with high methylation variability, and even for cluster 3 genes, there is also a trend that the negCRs are enriched at close downstream of TSS. It might give hypotheses that when the transcription machine moves into the gene body from TSS, these exist some mechanism that blocks this process and reflects on the changes of methylations.</p>
<p>H3K4me3 is a histone mark which is enriched at active TSS or promoters. Heatmap “H3K4me3_mean” shows strong enrichment of the mean signal over TSS for cluster 1 and cluster 2 genes with high expression. Such enrichment corresponds very well to the low TSS-methylation. Interestingly, strong positive correlation to expression dominates in cluster 1 and the signals are significantly higher in embryonic cells (heatmap “H3K4me3_diff”). The peak for the enrichment of correlation signals in cluster 1 (solid red line in annotation of heatmap “H3K4me3_corr”) is broader than the mean signals while it is very similar as the enrichment peak for negCRs. For cluster 2 genes, the positive correlated regions are enriched at downstream border of H3K4me3 peaks while directly at the H3K4me3 peaks shows negative correlation although the correlation signals are weak and signal difference is small. Surprisingly, strong positive correlations dominate cluster 3 although the mean signals selves are very weak.</p>
<p>H3K4me1 is an active mark enriched at enhancers and promoter flanking regions. Nevertheless, it shows negative correlation at the TSS (solid and dashed green lines in annotation of heatmap “H3K4me1_corr”), especially strong for cluster 1. The peak for the negative correlation enrichment correlates well with CGI and low TSS-methylation, however the signals selves are low at TSS (heatmap “H3K4me1_mean”). Flanking TSS is dominated by positive correlations and the signal difference is comparably big in cluster 1 (solid brown line in annotation of heatmap “H3K4me1_diff”).</p>
<p>H3K27ac is also an active mark enriched in both active enhancers and promoters, and it generally shows positive correlations to expression in all three clusters (heatmap “H3K27ac_corr”). Interestingly the mean signals are the strongest in cluster 2 and mature cells have significantly higher signal intensity than embryonic cells (dashed blue line in annotation of heatmap “H3K27ac_diff”). The peak for the correlation signal enrichment is comparably broader than other marks.</p>
<p>H3K27me3 is a repressive mark and it generally shows negative correlation around TSS at relatively low level, excluding cluster 1 where there are no dominant correlation patterns (heatmap “H3K27me3_corr”). The signals selves are lower and sparser compared to other marks.</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.2
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
## 
## attached base packages:
##  [1] parallel  stats4    grid      stats     graphics  grDevices utils     datasets  methods  
## [10] base     
## 
## other attached packages:
##  [1] genefilter_1.66.0      RColorBrewer_1.1-2     GetoptLong_0.1.7       EnrichedHeatmap_1.15.0
##  [5] GenomicRanges_1.36.1   GenomeInfoDb_1.20.0    IRanges_2.18.2         S4Vectors_0.22.1      
##  [9] BiocGenerics_0.30.0    ComplexHeatmap_2.1.0   circlize_0.4.8         knitr_1.25            
## [13] markdown_1.1          
## 
## loaded via a namespace (and not attached):
##  [1] shape_1.4.4            locfit_1.5-9.1         xfun_0.9               splines_3.6.1         
##  [5] lattice_0.20-38        vctrs_0.2.0            colorspace_1.4-1       htmltools_0.3.6       
##  [9] survival_2.44-1.1      blob_1.2.0             XML_3.98-1.20          rlang_0.4.0           
## [13] pillar_1.4.2           pkgdown_1.4.1          DBI_1.0.0              bit64_0.9-7           
## [17] matrixStats_0.55.0     GenomeInfoDbData_1.2.1 stringr_1.4.0          zlibbioc_1.30.0       
## [21] GlobalOptions_0.1.0    memoise_1.1.0          evaluate_0.14          Biobase_2.44.0        
## [25] AnnotationDbi_1.46.1   Rcpp_1.0.2             xtable_1.8-4           backports_1.1.4       
## [29] desc_1.2.0             annotate_1.62.0        XVector_0.24.0         fs_1.3.1              
## [33] bit_1.1-14             rjson_0.2.20           png_0.1-7              digest_0.6.21         
## [37] stringi_1.4.3          rprojroot_1.3-2        clue_0.3-57            tools_3.6.1           
## [41] bitops_1.0-6           magrittr_1.5           tibble_2.1.3           RCurl_1.95-4.12       
## [45] RSQLite_2.1.2          cluster_2.1.0          pkgconfig_2.0.3        zeallot_0.1.0         
## [49] crayon_1.3.4           Matrix_1.2-17          MASS_7.3-51.4          assertthat_0.2.1      
## [53] rmarkdown_1.15         R6_2.4.0               compiler_3.6.1</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#visualize-comprehensive-associations-in-roadmap-dataset">Visualize Comprehensive Associations in Roadmap dataset</a><ul class="nav nav-pills nav-stacked">
<li><a href="#general-overview">General overview</a></li>
      <li><a href="#association-between-gene-expression-and-methylation">Association between gene expression and methylation</a></li>
      <li><a href="#association-with-histone-modifications">Association with histone modifications</a></li>
      <li><a href="#normalize-cgi-to-gene-tss">normalize CGI to gene TSS</a></li>
      <li><a href="#order-and-group-genes">Order and group genes</a></li>
      <li><a href="#organize-heatmaps">Organize heatmaps</a></li>
      <li><a href="#interpretation">Interpretation</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
