<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • EnrichedHeatmap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EnrichedHeatmap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/EnrichedHeatmap.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/roadmap.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../articles/row_odering.html">UNKNOWN TITLE</a>
    </li>
    <li>
      <a href="../articles/visualize_categorical_signals_wrapper.html">UNKNOWN TITLE</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/EnrichedHeatmap">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/EnrichedHeatmap/blob/master/vignettes/visualize_categorical_signals_wrapper.Rmd"><code>vignettes/visualize_categorical_signals_wrapper.Rmd</code></a></small>
      <div class="hidden name"><code>visualize_categorical_signals_wrapper.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{2. Visualize Categorical Signals}
-->
<div id="visualize-categorical-signals" class="section level1">
<h1 class="hasAnchor">
<a href="#visualize-categorical-signals" class="anchor"></a>Visualize Categorical Signals</h1>
<p><strong>Author</strong>: Zuguang Gu ( <a href="mailto:z.gu@dkfz.de">z.gu@dkfz.de</a> )</p>
<p><strong>Date</strong>: 2018-11-10</p>
<hr>
<p>Normally genomic signals are represented as numeric (e.g. methylation from WGBS or histone modification intensity from ChIP-seq) or binary values (e.g. existance of CpG islands in a given position), however, genomic signals sometimes can also be represented as categorical or discrete values (mostly stored as characters). A very typical example is chromatin states segmentation by <a href="http://compbio.mit.edu/ChromHMM/">ChromHMM</a>. ChromHMM basically assigns a chromatin state (e.g. active transcription state or repressive transcription state) to a given window in the genome and the assignment of states in different windows are always mutually exclusive (which means one window can only have one state). In this vignette, We will demonstrate how to visualize the enrichment of various chromatin states around certain genomic targets and how to integerate with other epigenomic signals.</p>
<div id="general-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#general-examples" class="anchor"></a>General examples</h2>
<p>In following example, we use <a href="http://www.roadmapepigenomics.org/">Roadmap dataset</a> as the example dataset. First we show basic usages with using one sample (sample id: E003, embryonic stem cell, H1 cell line).</p>
<p>Load packages:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GenomicRanges)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(data.table)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(EnrichedHeatmap)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(circlize)</code></pre></div>
<p>The chromatin state segmentation is trained and applied from <a href="http://egg2.wustl.edu/roadmap/web_portal/chr_state_learning.html#core_15state">five histone modifications</a>. Following file in use is from <a href="http://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/E003_15_coreMarks_mnemonics.bed.gz" class="uri">http://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/E003_15_coreMarks_mnemonics.bed.gz</a>.</p>
<p>We convert to a <code>GRanges</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">states_bed =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/data.table/man/fread.html">fread</a></span>(<span class="st">"zcat ~/EnrichedHeatmap_test/E003_15_coreMarks_mnemonics.bed.gz"</span>)
states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(<span class="dt">seqnames =</span> states_bed[[<span class="dv">1</span>]], 
    <span class="dt">ranges =</span> <span class="kw">IRanges</span>(states_bed[[<span class="dv">2</span>]] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, states_bed[[<span class="dv">3</span>]]), 
    <span class="dt">states =</span> states_bed[[<span class="dv">4</span>]])
<span class="kw"><a href="https://rdrr.io/pkg/data.table/man/duplicated.html">unique</a></span>(states_bed[[<span class="dv">4</span>]])</code></pre></div>
<pre><code>##  [1] "15_Quies"    "9_Het"       "14_ReprPCWk" "13_ReprPC"   "10_TssBiv"   "7_Enh"      
##  [7] "1_TssA"      "11_BivFlnk"  "2_TssAFlnk"  "5_TxWk"      "4_Tx"        "8_ZNF/Rpts" 
## [13] "6_EnhG"      "12_EnhBiv"   "3_TxFlnk"</code></pre>
<p>In the 15 states, there are some states which are similar to each other such as <code>1_TssA</code> (active TSS) and <code>2_TssAFlnk</code> (flanking active TSS). To reduce the complexity of the analysis, we merge some of the similar states.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
    <span class="st">"1_TssA"</span>      =<span class="st"> "TssActive"</span>,
    <span class="st">"2_TssAFlnk"</span>  =<span class="st"> "TssActive"</span>,
    <span class="st">"3_TxFlnk"</span>    =<span class="st"> "Transcript"</span>,
    <span class="st">"4_Tx"</span>        =<span class="st"> "Transcript"</span>,
    <span class="st">"5_TxWk"</span>      =<span class="st"> "Transcript"</span>,
    <span class="st">"6_EnhG"</span>      =<span class="st"> "Enhancer"</span>,
    <span class="st">"7_Enh"</span>       =<span class="st"> "Enhancer"</span>,
    <span class="st">"8_ZNF/Rpts"</span>  =<span class="st"> "Heterochromatin"</span>,
    <span class="st">"9_Het"</span>       =<span class="st"> "Heterochromatin"</span>,
    <span class="st">"10_TssBiv"</span>   =<span class="st"> "TssBivalent"</span>,
    <span class="st">"11_BivFlnk"</span>  =<span class="st"> "TssBivalent"</span>,
    <span class="st">"12_EnhBiv"</span>   =<span class="st"> "Enhancer"</span>,
    <span class="st">"13_ReprPC"</span>   =<span class="st"> "Repressive"</span>,
    <span class="st">"14_ReprPCWk"</span> =<span class="st"> "Repressive"</span>,
    <span class="st">"15_Quies"</span>    =<span class="st"> "Quiescent"</span>
)
states<span class="op">$</span>states_simplified =<span class="st"> </span>map[states<span class="op">$</span>states]</code></pre></div>
<p>Also we set the colors for the 7 merged states.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">states_col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
    <span class="st">"TssActive"</span>       =<span class="st"> "Red"</span>,
    <span class="st">"Transcript"</span>      =<span class="st"> "Green"</span>,
    <span class="st">"Enhancer"</span>        =<span class="st"> "Yellow"</span>,
    <span class="st">"Heterochromatin"</span> =<span class="st"> "PaleTurquoise"</span>,
    <span class="st">"TssBivalent"</span>     =<span class="st"> "Orange"</span>,
    <span class="st">"Repressive"</span>      =<span class="st"> "Grey"</span>,
    <span class="st">"Quiescent"</span>       =<span class="st"> "black"</span>
)
states_name =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(states_col)
n_states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(states_col)</code></pre></div>
<p>In following, we demonstrate how to normalize the chromatin states to TSS. First we load the transcriptome and extract TSS regions. The transcriptome annotation is from <a href="http://egg2.wustl.edu/roadmap/data/byDataType/rna/annotations/gen10.long.gtf.gz" class="uri">http://egg2.wustl.edu/roadmap/data/byDataType/rna/annotations/gen10.long.gtf.gz</a> and we only use protein coding genes. The database file (the <code>sqlite</code> file) for transcriptome was generated by <strong>GenomicFeatures</strong> package (the <code>makeTxDbFromGFF()</code> function).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GenomicFeatures)
txdb =<span class="st"> </span><span class="kw">loadDb</span>(<span class="st">"~/EnrichedHeatmap_test/gencode19_protein_coding_txdb.sqlite"</span>)

g =<span class="st"> </span><span class="kw">genes</span>(txdb)
tss =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html">promoters</a></span>(g, <span class="dt">upstream =</span> <span class="dv">0</span>, <span class="dt">downstream =</span> <span class="dv">1</span>)</code></pre></div>
<p>To reduce the running time, here we only use chromosome 1. Normalizing categorical signals is basically as same as numeric signals. Here we only need to specify the column name for the categorical signals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tss_chr1 =<span class="st"> </span>tss[<span class="kw">seqnames</span>(tss) <span class="op">==</span><span class="st"> "chr1"</span>]
<span class="co"># column "states_simplified" is in character mode</span>
mat_states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, tss_chr1, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
mat_states</code></pre></div>
<pre><code>## Normalize states to tss_chr1:
##   Upstream 5000 bp (50 windows)
##   Downstream 5000 bp (50 windows)
##   Include target regions (width = 1)
##   2065 target regions
##   signal is categorical (7 levels)</code></pre>
<p>In the last line of the message, it clarifies it is categorical signals with 7 levels.</p>
<p>The implementation of normalizing categorical signals is actually very simple. Internally, <code>n_states</code> normalized matrices are generated where each one corresponds to one chromatin state and the value in each window is the fraction how much the window is overlapped to the state (with <code>w0</code> or <code>weighted</code> mean mode). Since the i^th row and the j^th column in all matrices correspond to a same window to a same target (here the TSS), if there are multiple states overlap to this same window, when summarizing from all states, the state with largest overlap fraction is assigned to this window.</p>
<p>Since the <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix()</a></code> is called <code>n_states</code> times internally, it will be a little bit slow to normalize to categorical signals.</p>
<p>There are some special visualizations designed for categorical signals where each state is summarized and visualized separatedly in the top annotation. Here <code>states_col</code> must be a named vector where the names should correspond to the levels of the categorical signals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col)</code></pre></div>
<div class="figure">
<img src="image/categorical_default-1.png" alt="plot of chunk categorical_default"><p class="caption">plot of chunk categorical_default</p>
</div>
<p>You might feel the row ordering is a little bit in a mess. Although the signals are categorical, internally, there are coded as integer numbers. Just similar as how <strong>factor</strong> in R is stored, each categorical level corresponds to an integer number. If the signal column is simply a character vector, the assignment of integers are based on the natural ordering of this character vector. Thus the numeric coding for signals in <code>states</code> object is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">states =</span> <span class="kw"><a href="https://rdrr.io/pkg/data.table/man/duplicated.html">unique</a></span>(states<span class="op">$</span>states_simplified), <span class="dt">value =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>)</code></pre></div>
<pre><code>##            states value
## 1       Quiescent     1
## 2 Heterochromatin     2
## 3      Repressive     3
## 4     TssBivalent     4
## 5        Enhancer     5
## 6       TssActive     6
## 7      Transcript     7</code></pre>
<p>Zero is assigned to a window if none of the states overlap to it and it is drawn with white.</p>
<p>The numeric coding can be controlled by setting the signals as a factor variable and the level order of the factor controls the corresponding coding. In following code, we convert <code>states_simplified</code> as a factor with specifying the level order.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">states<span class="op">$</span>states_simplified =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(states<span class="op">$</span>states_simplified, <span class="dt">levels =</span> states_name)
<span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">states =</span> <span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(states<span class="op">$</span>states_simplified), <span class="dt">value =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>)</code></pre></div>
<pre><code>##            states value
## 1       TssActive     1
## 2      Transcript     2
## 3        Enhancer     3
## 4 Heterochromatin     4
## 5     TssBivalent     5
## 6      Repressive     6
## 7       Quiescent     7</code></pre>
<p>Note here the order of <code>states_name</code> also reflects the closeness of different states. With the order defined above, <code>TssActive</code> is closer to <code>Transcript</code> and <code>Repressive</code> is closer to <code>Quiescent</code>.</p>
<p>In following left plot, the TSS with active states now are far from the TSS with repressive states. In the right plot we change the level of <code>states_simplified</code> and you can compare to the left one.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, tss_chr1, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col)
<span class="co"># shuffle levels for states_simplified</span>
states<span class="op">$</span>states_simplified =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(states<span class="op">$</span>states_simplified, 
    <span class="dt">levels =</span> states_name[<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">6</span>, <span class="dv">2</span>, <span class="dv">7</span>, <span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">5</span>)])
mat_states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, tss_chr1, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col, 
    <span class="dt">row_title =</span> <span class="st">"states_name[c(1, 6, 2, 7, 4, 3, 5)]"</span>)</code></pre></div>
<div class="figure">
<img src="image/unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10"><p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>We suggest to apply hierarchical clustering to reorder rows. Here the numeric coding is espeically important because it affects the calculation of distance between rows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we change the level back</span>
states<span class="op">$</span>states_simplified =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(states<span class="op">$</span>states_simplified, <span class="dt">levels =</span> states_name)
mat_states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, tss_chr1, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>)</code></pre></div>
<div class="figure">
<img src="image/categorical_clustered-1.png" alt="plot of chunk categorical_clustered"><p class="caption">plot of chunk categorical_clustered</p>
</div>
<p>Since the normalized matrix is numeric, k-means clustering can also be applied.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>, <span class="dt">row_km =</span> <span class="dv">2</span>)</code></pre></div>
<div class="figure">
<img src="image/categorical_kmeans-1.png" alt="plot of chunk categorical_kmeans"><p class="caption">plot of chunk categorical_kmeans</p>
</div>
<p>After closely looking at the heatmap, we found the TSS states are consistently enriched around TSS while in the flanking regions, the states are a little bit diverse. Actually this suggests that in order to enhance the pattern of the TSS-related states, we can apply k-means clustering only for the states near TSS regions.</p>
<p>In following we only cluster the sub-matrix which is upstrean and downstream 1kb of TSS (The default extension of TSS is 5kb upstream and downstream, the number of columns in the normalized matrix is 100, thus from 40^th column to 60^th column are the states in 1kb upstream and downstream of TSS).</p>
<p>Of couse we need to calculate this partitioning in advance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(mat_states[, <span class="dv">40</span><span class="op">:</span><span class="dv">60</span>], <span class="dt">centers =</span> <span class="dv">2</span>)<span class="op">$</span>cluster
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>, 
    <span class="dt">row_split =</span> split)</code></pre></div>
<div class="figure">
<img src="image/categorical_partial_kmeans-1.png" alt="plot of chunk categorical_partial_kmeans"><p class="caption">plot of chunk categorical_partial_kmeans</p>
</div>
<p>Now it is quite nice to see genes with active TSS are all clustered in cluster 2 while in cluster 1, genes are either have no function or have bivalent TSS function.</p>
<p>Similarly, we can visualize how the chromatin states enriched at gene bodies. Since the gene bodies have unequal widths, we add an additional point plot to show the width of genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">g_chr1 =<span class="st"> </span>g[<span class="kw">seqnames</span>(g) <span class="op">==</span><span class="st"> "chr1"</span>]
mat_states_<span class="dv">2</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, g_chr1, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states_<span class="dv">2</span>, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>) <span class="op">+</span>
<span class="kw">rowAnnotation</span>(<span class="dt">gene_len =</span> <span class="kw">anno_points</span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="kw">width</span>(g_chr1) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">size =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">"mm"</span>),
    <span class="dt">axis =</span> <span class="ot">TRUE</span>, <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">at =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>), 
        <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"100bp"</span>, <span class="st">"1kb"</span>, <span class="st">"10kb"</span>, <span class="st">"100kb"</span>, <span class="st">"1mb"</span>)), 
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">4</span>, <span class="st">"cm"</span>)))</code></pre></div>
<div class="figure">
<img src="image/categorical_gene_body-1.png" alt="plot of chunk categorical_gene_body"><p class="caption">plot of chunk categorical_gene_body</p>
</div>
<p>Next we add methylation around TSS as well as expression for the corresponding genes. Expression data is from <a href="http://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/57epigenomes.RPKM.pc.gz" class="uri">http://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/57epigenomes.RPKM.pc.gz</a> and methylation data is from <a href="http://egg2.wustl.edu/roadmap/data/byDataType/dnamethylation/WGBS/FractionalMethylation.tar.gz" class="uri">http://egg2.wustl.edu/roadmap/data/byDataType/dnamethylation/WGBS/FractionalMethylation.tar.gz</a></p>
<p>In following, the raw methylation data has been smoothed by <strong>bsseq</strong> package. The rds file was generated by the old version of <strong>bsseq</strong> package that is why we used a strange way to extract the positions of CpG sites as well as smoothed methylation data which can be replaced by using <code><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html">granges()</a></code> and <code>getMeth()</code> functions if the smoothing is applied with newest version of <strong>bsseq</strong> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"~/EnrichedHeatmap_test/57epigenomes.RPKM.pc.gz"</span>, <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)
expr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/data.table/man/as.matrix.html">as.matrix</a></span>(expr)
obj =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="st">"~/EnrichedHeatmap_test/chr1_roadmap_merged_bsseq.rds"</span>)
meth =<span class="st"> </span>obj<span class="op">@</span>rowData
meth_mat =<span class="st"> </span>obj<span class="op">@</span><span class="kw">trans</span>(obj<span class="op">@</span>assays<span class="op">$</span>data<span class="op">$</span>coef)
<span class="kw">mcols</span>(meth) =<span class="st"> </span>meth_mat</code></pre></div>
<p>We take the genes which exist both in <code>tss_chr1</code> and <code>expr</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(tss_chr1) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"</span><span class="ch">\\</span><span class="st">.</span><span class="ch">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(tss_chr1))
cn =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/setops-methods.html">intersect</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(tss_chr1), <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(expr))
tss_chr1 =<span class="st"> </span>tss_chr1[cn]
expr =<span class="st"> </span>expr[cn, ]</code></pre></div>
<p>We normalize chromatin states as well methylation to TSS on chromosome 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_states =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, tss_chr1, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
mat_meth =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, tss_chr1, <span class="dt">value_column =</span> <span class="st">"E003"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">smooth =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Similarly, k-means is applied on chromatin states 1kb upstream and downstream of TSS. Note since the rows are split into two sub-cluster, we add <code>anno_summary()</code> to the expression matrix to show the distribution of expression in the two sub-clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(mat_states[, <span class="dv">40</span><span class="op">:</span><span class="dv">60</span>], <span class="dt">centers =</span> <span class="dv">2</span>)<span class="op">$</span>cluster
meth_col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))
ht_list =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states, <span class="dt">name =</span> <span class="st">"states"</span>, <span class="dt">col =</span> states_col, <span class="dt">cluster_rows =</span> <span class="ot">TRUE</span>, 
    <span class="dt">row_split =</span> split, 
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)))) <span class="op">+</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_meth, <span class="dt">name =</span> <span class="st">"meth"</span>, <span class="dt">col =</span> meth_col_fun,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)))) <span class="op">+</span>
<span class="kw">Heatmap</span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(expr[<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(tss_chr1), <span class="st">"E003"</span>] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">name =</span> <span class="st">"expr"</span>, 
    <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">"cm"</span>),
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">summary =</span> <span class="kw">anno_summary</span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>),
        <span class="dt">outline =</span> <span class="ot">FALSE</span>, <span class="dt">axis_param =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">side =</span> <span class="st">"right"</span>))))
<span class="kw">draw</span>(ht_list, <span class="dt">ht_gap =</span> <span class="kw">unit</span>(<span class="dv">8</span>, <span class="st">"mm"</span>))</code></pre></div>
<div class="figure">
<img src="image/categorical_with_meth-1.png" alt="plot of chunk categorical_with_meth"><p class="caption">plot of chunk categorical_with_meth</p>
</div>
<p>With corresponding to more epigenomic signals, we can see promoters with active TSS states are always associated with low methylation and high gene expression while promoters annotated as bivalent states or repressed are generally more lowly expressed.</p>
</div>
<div id="bivalent-tss" class="section level2">
<h2 class="hasAnchor">
<a href="#bivalent-tss" class="anchor"></a>Bivalent TSS</h2>
<p>It is believed that in embryonic stem cell, there are a huge number of genes showing bivalency states on promoters. They are called bivalent states because both H3K4me3 which is a histone mark for active transcription and H3K27me3 which is for repressive transcription exist. The loss/gain of active histone mark or repressive histone mark which transite into mature cells is essential for tissue development and differentiation.</p>
<p>To demonstate this, we anlaysis such transitions for one mature tissue (lung tissue, roadmap sample id: E096).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">states_bed =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/data.table/man/fread.html">fread</a></span>(<span class="st">"zcat ~/EnrichedHeatmap_test/E096_15_coreMarks_mnemonics.bed.gz"</span>)
states_lung =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(<span class="dt">seqnames =</span> states_bed[[<span class="dv">1</span>]], <span class="dt">ranges =</span> <span class="kw">IRanges</span>(states_bed[[<span class="dv">2</span>]] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, states_bed[[<span class="dv">3</span>]]), 
    <span class="dt">states =</span> states_bed[[<span class="dv">4</span>]])
states_lung<span class="op">$</span>states_simplified =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(map[states_lung<span class="op">$</span>states], <span class="dt">levels =</span> states_name)</code></pre></div>
<p>ChromHMM is performed with 200bp window, thus we split whole genome by 200bp window and assign corresponding states to each window, both for ESC and lung.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">window =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/makeWindows.html">makeWindows</a></span>(states, <span class="dt">w =</span> <span class="dv">200</span>)
mtch =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/data.table/man/as.matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">findOverlaps</a></span>(window, states))
window<span class="op">$</span>ESC_states =<span class="st"> </span>states<span class="op">$</span>states_simplified[mtch[, <span class="dv">2</span>]]
mtch =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/data.table/man/as.matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">findOverlaps</a></span>(window, states_lung))
window<span class="op">$</span>lung_states =<span class="st"> </span>states_lung<span class="op">$</span>states_simplified[mtch[, <span class="dv">2</span>]]</code></pre></div>
<p>We only use the windows which are annotated with <code>TssBivalent</code> state either in ESC or in lung.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">window =<span class="st"> </span>window[window<span class="op">$</span>ESC_states <span class="op">==</span><span class="st"> "TssBivalent"</span> <span class="op">|</span><span class="st"> </span>window<span class="op">$</span>lung_states <span class="op">==</span><span class="st"> "TssBivalent"</span>]</code></pre></div>
<p>We first make a Chord diagram to show how the transsition happens.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">transition_mat =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="kw">mcols</span>(window)[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ESC_states"</span>, <span class="st">"lung_states"</span>)]) <span class="op">*</span><span class="st"> </span><span class="dv">200</span>
<span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(transition_mat) =<span class="st"> "matrix"</span>
transition_mat =<span class="st"> </span>transition_mat[states_name, states_name]
transition_mat</code></pre></div>
<pre><code>##                  lung_states
## ESC_states        TssActive Transcript Enhancer Heterochromatin TssBivalent Repressive Quiescent
##   TssActive               0          0        0               0      136200          0         0
##   Transcript              0          0        0               0        8200          0         0
##   Enhancer                0          0        0               0       38800          0         0
##   Heterochromatin         0          0        0               0       55800          0         0
##   TssBivalent       4485600     665400  3711800          653000      406000    1799200    766200
##   Repressive              0          0        0               0       24600          0         0
##   Quiescent               0          0        0               0       23800          0         0</code></pre>
<p>The values in the transition matrix mean how many base pairs change from one chromatin state to the other state.</p>
<p>Rows and columns are from different cells, we add prefix to row and column names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(transition_mat) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"ESC_"</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(transition_mat))
<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(transition_mat) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"lung_"</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(transition_mat))</code></pre></div>
<p>Now we make the Chord diagram.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grid.col =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(states_col, states_col)
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(grid.col) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(transition_mat), <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(transition_mat))
<span class="kw"><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html">chordDiagram</a></span>(transition_mat, <span class="dt">grid.col =</span> grid.col, <span class="dt">annotationTrack =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"grid"</span>, <span class="st">"axis"</span>),
    <span class="dt">directional =</span> <span class="ot">TRUE</span>, <span class="dt">gap.degree =</span> <span class="dv">5</span>)
<span class="kw"><a href="https://rdrr.io/pkg/circlize/man/circos.clear.html">circos.clear</a></span>()
<span class="kw"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="fl">0.5</span>, <span class="op">-</span><span class="dv">1</span>, <span class="st">"ESC"</span>)
<span class="kw"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="st">"lung"</span>)
<span class="kw"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"left"</span>, <span class="dt">pch =</span> <span class="dv">15</span>, <span class="dt">col =</span> states_col, <span class="dt">legend =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(states_col))</code></pre></div>
<div class="figure">
<img src="image/tssbiv_chord_diagram-1.png" alt="plot of chunk tssbiv_chord_diagram"><p class="caption">plot of chunk tssbiv_chord_diagram</p>
</div>
<p>As we can see from the Chord diagram, a lot of regions with bivalent TSS states (orange track at bottom) have been tansite to active (red track on top) or repressive states (grey track on top) in lung.</p>
<p>To get more deep of how the transistion looks like, we can make enriched heatmaps to see the associations between different epigenomic signals.</p>
<p>Since <code>TssBivalent</code> states basically are states for TSS related regions, we only keep genes for which in 1kb upstream and downstream of TSS there must be a window with TssBivalent state. Again, we only use chromosome 1 as demonstration.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_bivtss =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states[states<span class="op">$</span>states_simplified <span class="op">==</span><span class="st"> "TssBivalent"</span>], tss_chr1)
l =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(mat_bivtss[, <span class="dv">40</span><span class="op">:</span><span class="dv">60</span>]) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span> <span class="co"># 1kb upstream and downstream</span>
tss_biv =<span class="st"> </span>tss_chr1[l]
tss_biv</code></pre></div>
<pre><code>## GRanges object with 443 ranges and 1 metadata column:
##                   seqnames                 ranges strand |           gene_id
##                      &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;character&gt;
##   ENSG00000000938     chr1 [ 27961788,  27961788]      - | ENSG00000000938.8
##   ENSG00000006555     chr1 [ 55266940,  55266940]      - | ENSG00000006555.6
##   ENSG00000007968     chr1 [ 23857712,  23857712]      - | ENSG00000007968.6
##   ENSG00000008118     chr1 [209757062, 209757062]      + | ENSG00000008118.5
##   ENSG00000009709     chr1 [ 18957500,  18957500]      + | ENSG00000009709.7
##               ...      ...                    ...    ... .               ...
##   ENSG00000243749     chr1 [ 35450954,  35450954]      - | ENSG00000243749.1
##   ENSG00000248485     chr1 [161228517, 161228517]      + | ENSG00000248485.1
##   ENSG00000249087     chr1 [ 23695711,  23695711]      + | ENSG00000249087.3
##   ENSG00000251246     chr1 [155036224, 155036224]      + | ENSG00000251246.1
##   ENSG00000253304     chr1 [ 29450447,  29450447]      - | ENSG00000253304.1
##   -------
##   seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We normalize chromatin states and methylation to <code>tss_biv</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_states_ESC =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states, tss_biv, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)
mat_states_lung =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(states_lung, tss_biv, <span class="dt">value_column =</span> <span class="st">"states_simplified"</span>)

mat_meth_ESC =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, tss_biv, <span class="dt">value_column =</span> <span class="st">"E003"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">smooth =</span> <span class="ot">TRUE</span>)
mat_meth_lung =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span>(meth, tss_biv, <span class="dt">value_column =</span> <span class="st">"E096"</span>, <span class="dt">mean_mode =</span> <span class="st">"absolute"</span>,
    <span class="dt">smooth =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Normally, methylation changes happen at the border of low methylation region. To enhance the effect of methylation change, we directly visualize the methyaltion difference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_meth_diff =<span class="st"> </span>mat_meth_ESC <span class="op">-</span><span class="st"> </span>mat_meth_lung
meth_diff_col_fun =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="fl">0.25</span>, <span class="dv">0</span>, <span class="fl">0.25</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#3794bf"</span>, <span class="st">"#FFFFFF"</span>, <span class="st">"#df8640"</span>))</code></pre></div>
<p>Now we construct the heatmap list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states_ESC, <span class="dt">name =</span> <span class="st">"states_ESC"</span>, <span class="dt">col =</span> states_col,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>))),
    <span class="dt">column_title =</span> <span class="st">"States ESC"</span>) <span class="op">+</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_states_lung, <span class="dt">name =</span> <span class="st">"states_lung"</span>, <span class="dt">col =</span> states_col,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>))),
    <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>, <span class="dt">column_title =</span> <span class="st">"States lung"</span>)

ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_meth_ESC, <span class="dt">name =</span> <span class="st">"meth_ESC"</span>, <span class="dt">col =</span> meth_col_fun,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>))),
    <span class="dt">column_title =</span> <span class="st">"Meth ESC"</span>) <span class="op">+</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_meth_lung, <span class="dt">name =</span> <span class="st">"meth_lung"</span>, <span class="dt">col =</span> meth_col_fun,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), 
        <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>))),
    <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>, <span class="dt">column_title =</span> <span class="st">"Meth lung"</span>) <span class="op">+</span>
<span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_meth_diff, <span class="dt">name =</span> <span class="st">"meth_diff"</span>, <span class="dt">col =</span> meth_diff_col_fun,
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, 
        <span class="dt">pos_col =</span> <span class="st">"#df8640"</span>, <span class="dt">neg_col =</span> <span class="st">"#3794bf"</span>))),
    <span class="dt">column_title =</span> <span class="st">"Meth ESC - lung"</span>)</code></pre></div>
<p>Actually we can discretize the numeric <code>mat_meth_diff</code>. Here we assign <code>hyper</code> to windows with methylation difference larger than 0.2 and <code>hypo</code> to windows with methylation difference less than -0.2. Here <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/discretize.html">discretize()</a></code> converts a matrix with continuous signals to categorical signals by prividing a list of intervals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_meth_diff_discrete =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/discretize.html">discretize</a></span>(mat_meth_diff,
    <span class="dt">rule =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
        <span class="st">"hypo"</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="ot">Inf</span>, <span class="op">-</span><span class="fl">0.2</span>),
        <span class="st">"hyper"</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.2</span>, <span class="ot">Inf</span>)
    )
)
mat_meth_diff_discrete</code></pre></div>
<pre><code>## Normalize  to tss_biv:
##   Upstream 5000 bp (50 windows)
##   Downstream 5000 bp (50 windows)
##   Include target regions (width = 1)
##   443 target regions
##   signal is categorical (2 levels)</code></pre>
<p>We continue to add more heatmaps.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span>(mat_meth_diff_discrete, <span class="dt">name =</span> <span class="st">"meth_diff_discrete"</span>, 
    <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hyper"</span> =<span class="st"> "#df8640"</span>, <span class="dt">hypo =</span> <span class="st">"#3794bf"</span>),
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">enrich =</span> <span class="kw"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>))))</code></pre></div>
<p>Finally the heatmap for gene expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(expr[<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(tss_biv), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"E003"</span>, <span class="st">"E096"</span>)] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw">Heatmap</span>(e, <span class="dt">name =</span> <span class="st">"expr"</span>, 
    <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">10</span>, <span class="st">"mm"</span>), <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>The row ordering for all heatmaps is from hierarchical clustering on the merged matrix from states in ESC and lung, 1kb upstream and downstream of TSS.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">row_order =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(mat_states_ESC[, <span class="dv">40</span><span class="op">:</span><span class="dv">60</span>], mat_states_lung[, <span class="dv">40</span><span class="op">:</span><span class="dv">60</span>])))<span class="op">$</span>order
split =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(e[, <span class="st">"E096"</span>] <span class="op">&gt;</span><span class="st"> </span>e[, <span class="st">"E003"</span>], <span class="st">"activation"</span>, <span class="st">"repression"</span>)
<span class="kw">draw</span>(ht_list, <span class="dt">row_order =</span> row_order, <span class="dt">row_split =</span> split)</code></pre></div>
<div class="figure">
<img src="image/categorical_with_lung-1.png" alt="plot of chunk categorical_with_lung"><p class="caption">plot of chunk categorical_with_lung</p>
</div>
<p>What conclusion can you make from the plot? I can only say the regulation pattern is really complicated there :).</p>
</div>
<div id="other-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#other-examples" class="anchor"></a>Other examples</h2>
<p>There are some other examples where the genomic signals are categorical.</p>
<ol style="list-style-type: decimal">
<li>repeats where different repeat family can be different categories,</li>
<li>genome segmentation based on methylation. E.g. hign methylated regions (HMRs), partially methylated regions (PMDs), low methylated regions (LMRs), unmethylated regions (UMRs),</li>
<li>for gene-related regions, genic annotation is also categorical.</li>
</ol>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre></div>
<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: openSUSE 13.1 (Bottle) (x86_64)
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8       
##  [4] LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      parallel  stats4    stats     graphics  grDevices utils     datasets  methods  
## [10] base     
## 
## other attached packages:
##  [1] GenomicFeatures_1.26.4 AnnotationDbi_1.36.2   Biobase_2.34.0         circlize_0.4.5        
##  [5] EnrichedHeatmap_1.13.1 ComplexHeatmap_1.99.0  data.table_1.10.4      GenomicRanges_1.26.4  
##  [9] GenomeInfoDb_1.10.3    IRanges_2.8.2          S4Vectors_0.12.2       BiocGenerics_0.20.0   
## [13] markdown_0.8           knitr_1.20             colorout_1.2-0        
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.18               highr_0.7                  RColorBrewer_1.1-2        
##  [4] XVector_0.14.1             bitops_1.0-6               tools_3.3.1               
##  [7] zlibbioc_1.20.0            biomaRt_2.30.0             evaluate_0.11             
## [10] RSQLite_1.0.0              lattice_0.20-34            png_0.1-7                 
## [13] Matrix_1.2-7.1             DBI_1.0.0                  rtracklayer_1.34.2        
## [16] stringr_1.3.1              Biostrings_2.42.1          GlobalOptions_0.1.0       
## [19] locfit_1.5-9.1             GetoptLong_0.1.7           BiocParallel_1.8.2        
## [22] XML_3.98-1.16              magrittr_1.5               GenomicAlignments_1.10.1  
## [25] Rsamtools_1.26.2           matrixStats_0.54.0         SummarizedExperiment_1.4.0
## [28] shape_1.4.4                colorspace_1.3-2           stringi_1.2.4             
## [31] RCurl_1.95-4.8             rjson_0.2.20</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#visualize-categorical-signals">Visualize Categorical Signals</a><ul class="nav nav-pills nav-stacked">
<li><a href="#general-examples">General examples</a></li>
      <li><a href="#bivalent-tss">Bivalent TSS</a></li>
      <li><a href="#other-examples">Other examples</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
